

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>nasp.nasp_objects &mdash; Northern Arizona SNP Pipeline 0.9.8 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="Northern Arizona SNP Pipeline 0.9.8 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="../../index.html" class="fa fa-home"> Northern Arizona SNP Pipeline</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">The Northern Arizona SNP Pipeline (NASP)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#usage">Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#license">License</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#contact">Contact</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#install">Install</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#update">Update</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#uninstall">Uninstall</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#required-dependencies">Required Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#optional-dependencies">Optional Dependencies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../usage.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#nasp">nasp</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#format-fasta">format_fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#find-duplicates">find_duplicates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#vcf-to-matrix">vcf_to_matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#filter-matrix-by-coord-py">filter_matrix_by_coord.py</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#get-started">Get Started!</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#pull-request-guidelines">Pull Request Guidelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#tips">Tips</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../authors.html#development-lead">Development Lead</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../authors.html#contributors">Contributors</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html#id1">0.9.6 (2014-06-05)</a></li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../index.html">Northern Arizona SNP Pipeline</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../index.html">Module code</a> &raquo;</li>
      
    <li>nasp.nasp_objects</li>
      <li class="wy-breadcrumbs-aside">
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <h1>Source code for nasp.nasp_objects</h1><div class="highlight"><pre>
<span class="c">#!/usr/bin/env python3</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s">&quot;David Smith&quot;</span>
<span class="n">__version__</span> <span class="o">=</span> <span class="s">&quot;0.9.8&quot;</span>
<span class="n">__email__</span> <span class="o">=</span> <span class="s">&quot;dsmith@tgen.org&quot;</span>

<span class="kn">import</span> <span class="nn">logging</span>


<div class="viewcode-block" id="GenomeStatus"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus">[docs]</a><span class="k">class</span> <span class="nc">GenomeStatus</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Contains and manipulates any generic data that is per-contig-position.</span>
<span class="sd">    This could be any single type of data, like actual bases, filter data,</span>
<span class="sd">    depth information, insertion data, pileups, etc.</span>
<span class="sd">    In the perl version, this object was originally a hash of lists, and</span>
<span class="sd">    converted to a hash of strings for performance.  In this version, it was</span>
<span class="sd">    originally a dictionary of strings, and converted to a dictionary of</span>
<span class="sd">    lists for performance and flexibility.</span>
<span class="sd">    Whether single characters, numbers, boolean values, or a mixture of data</span>
<span class="sd">    types are used does not seem to affect memory and performance.</span>
<span class="sd">    Storing lists per-contig-position with this class is a complicated</span>
<span class="sd">    affair, as several of the manipulation functions assume you mean to</span>
<span class="sd">    manipulate a continuous range of positions instead of a single position</span>
<span class="sd">    when you do that.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    The conventions used for what data is stored are as follows:</span>
<span class="sd">    Genomes:</span>
<span class="sd">        A, C, G, T, U, a, c, g, t, u:  The respective call.</span>
<span class="sd">        N, n:  Called &quot;N&quot; according to upstream analysis tools.</span>
<span class="sd">        X:  Not called by upstream analysis tools.</span>
<span class="sd">        . or empty string:  A deletion relative to reference.</span>
<span class="sd">        String of length &gt;1:  An insertion relative to reference.</span>
<span class="sd">        Any other single letter:  A degeneracy.</span>
<span class="sd">        !:  An algorithmic failure.</span>
<span class="sd">    Duplicate region data:</span>
<span class="sd">        0:  Position not in a region that is duplicated within the reference.</span>
<span class="sd">        1:  Position is in a region that is duplicated.</span>
<span class="sd">        -:  Duplicate checking at this position was skipped by the user.</span>
<span class="sd">        !:  An algorithmic failure.</span>
<span class="sd">    Filters:</span>
<span class="sd">        Y:  This position passed its filter.</span>
<span class="sd">        N:  This position failed its filter.</span>
<span class="sd">        ?:  The filter could not be checked, and so the position is assumed</span>
<span class="sd">            to have failed.</span>
<span class="sd">        -:  The filter was not applicable, or skipped, or could not be checked</span>
<span class="sd">            for a known reason, and so is assumed to have passed.</span>
<span class="sd">        !:  An algorithmic failure.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c"># Arrays are zero-indexed, genome positions are one-indexed. Off-by-one errors? Never heard of &#39;em.</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attributes:</span>
<span class="sd">            _status_data: The dictionary of lists that stores the actual genome.</span>
<span class="sd">            data. The keys of the dictionary are the contig names.  The lists</span>
<span class="sd">            correspond to the position data on that contig.  Genome position is</span>
<span class="sd">            list position + 1.</span>
<span class="sd">            _current_contig: Tracks the most recently-referenced contig, for</span>
<span class="sd">            convenience EG reading in fastas line-by-line.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_current_contig</span> <span class="o">=</span> <span class="bp">None</span>

    <span class="c"># NOTE(jtravis): Does not raise exception if contig name is None or the empty string as documented</span>
<div class="viewcode-block" id="GenomeStatus.add_contig"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.add_contig">[docs]</a>    <span class="k">def</span> <span class="nf">add_contig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Defines a new empty contig in the genome.</span>
<span class="sd">        By default, if an unrecognized contig is encountered, a new empty</span>
<span class="sd">        contig will be created and then acted upon.</span>
<span class="sd">        Otherwise, add_contig must be called on a new contig first, or an</span>
<span class="sd">        InvalidContigName will be thrown.</span>

<span class="sd">        Args:</span>
<span class="sd">            contig_name (str): Unique contig description.</span>

<span class="sd">        Raises:</span>
<span class="sd">            InvalidContigName: If contig_name is undefined.</span>

<span class="sd">        Example:</span>
<span class="sd">            GenomeStatus.add_contig(&#39;500WT1_test&#39;)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">contig_name</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_current_contig</span> <span class="o">=</span> <span class="n">contig_name</span>

    <span class="c"># NOTE(jtravis): unused parameter create_contig</span></div>
<div class="viewcode-block" id="GenomeStatus.set_current_contig"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.set_current_contig">[docs]</a>    <span class="k">def</span> <span class="nf">set_current_contig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">,</span> <span class="n">create_contig</span><span class="o">=</span><span class="bp">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets the most-recently-referenced contig without actually performing</span>
<span class="sd">        any action on the data.</span>
<span class="sd">        Can be called to return the current contig without changing it if</span>
<span class="sd">        given a contig_name of None.</span>
<span class="sd">        Will create the contig if it has not been encountered yet by</span>
<span class="sd">        default, or throw an InvalidContigName otherwise.</span>

<span class="sd">        Args:</span>
<span class="sd">            contig_name (str): Unique contig description or None to query the current contig name.</span>
<span class="sd">            create_contig (bool): If True and the contig does not exist, an empty contig will be created.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Name of the last accessed contig or None.</span>

<span class="sd">        Raises:</span>
<span class="sd">            InvalidContigName: If create_contig is False and the contig does not exist.</span>

<span class="sd">        Example:</span>
<span class="sd">            GenomeStatus.set_current_contig(&#39;500WT1_test&#39;)</span>
<span class="sd">            GenomeStatus.set_current_contig(None)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">contig_name</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">contig_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_contig</span>
        <span class="k">elif</span> <span class="n">contig_name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_current_contig</span> <span class="o">=</span> <span class="n">contig_name</span>
        <span class="k">elif</span> <span class="n">create_contig</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">InvalidContigName</span><span class="p">(</span><span class="n">contig_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_contigs</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">contig_name</span>
</div>
<div class="viewcode-block" id="GenomeStatus.get_contigs"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.get_contigs">[docs]</a>    <span class="k">def</span> <span class="nf">get_contigs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns:</span>
<span class="sd">            list: Sorted list of contig names.</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; GenomeStatus.get_contigs()</span>
<span class="sd">            [&#39;500WT1_test&#39;, ...]</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
</div>
<div class="viewcode-block" id="GenomeStatus.append_contig"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.append_contig">[docs]</a>    <span class="k">def</span> <span class="nf">append_contig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome_data</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Places the passed-in data at the position following the last</span>
<span class="sd">        defined position on the contig.  If passed a list, will give each</span>
<span class="sd">        item in the list its own position.</span>

<span class="sd">        Args:</span>
<span class="sd">            genome_data (list or char): List of nucleotide symbols.</span>
<span class="sd">            contig_name (str): Unique contig description.</span>

<span class="sd">        Examples:</span>
<span class="sd">            append_contig([&#39;G&#39;, &#39;A&#39;, &#39;T&#39;, &#39;C&#39;, &#39;N&#39;], None)</span>
<span class="sd">            append_contig([&#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;, &#39;0&#39;], None)</span>
<span class="sd">            append_contig( &#39;A&#39;, &#39;(Reference)&#39; )</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">contig_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">set_current_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">genome_data</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="GenomeStatus.extend_contig"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.extend_contig">[docs]</a>    <span class="k">def</span> <span class="nf">extend_contig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_length</span><span class="p">,</span> <span class="n">missing_range_filler</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Ensures the contig is at least new_length positions long</span>

<span class="sd">        Args:</span>
<span class="sd">            new_length (int): Minimum contig length.</span>
<span class="sd">            missing_range_filler (str): Placeholder character for undefined areas at the end of the contig.</span>
<span class="sd">            contig_name (str): Unique contig description.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">contig_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">set_current_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">])</span> <span class="o">&lt;</span> <span class="n">new_length</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">]</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span>
                <span class="p">[</span><span class="n">missing_range_filler</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span> <span class="n">new_length</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">])</span> <span class="p">))</span>
</div>
<div class="viewcode-block" id="GenomeStatus.set_value"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.set_value">[docs]</a>    <span class="k">def</span> <span class="nf">set_value</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_data</span><span class="p">,</span> <span class="n">position_number</span><span class="p">,</span> <span class="n">missing_range_filler</span><span class="o">=</span><span class="s">&quot;!&quot;</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Sets the value at position_number on the contig.</span>
<span class="sd">        If passed a list, will change the continuous range of positions</span>
<span class="sd">        starting at position_number, one position per list item.</span>
<span class="sd">        Will extend the contig with missing_range_filler filling undefined</span>
<span class="sd">        values if the position to set is beyond the end of the contig.</span>

<span class="sd">        Args:</span>
<span class="sd">            new_data (str or list): Single or list of nucleotide symbols.</span>
<span class="sd">            position_number (int): 1-indexed contig position number.</span>
<span class="sd">            missing_range_filler (str): Filler for undefined regions before the set value. Modifies the data.</span>
<span class="sd">            contig_name (str): Unique contig description</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">contig_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">set_current_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">extend_contig</span><span class="p">(</span><span class="n">position_number</span><span class="p">,</span> <span class="n">missing_range_filler</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">new_data</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">][</span><span class="n">position_number</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span><span class="n">position_number</span> <span class="o">-</span> <span class="mi">1</span> <span class="o">+</span> <span class="nb">len</span><span class="p">(</span><span class="n">new_data</span><span class="p">)]</span> <span class="o">=</span> <span class="n">new_data</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">][</span><span class="n">position_number</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_data</span>
</div>
<div class="viewcode-block" id="GenomeStatus.get_value"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.get_value">[docs]</a>    <span class="k">def</span> <span class="nf">get_value</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">first_position</span><span class="p">,</span> <span class="n">last_position</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">filler_value</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            contig_name (str): Unique contig description.</span>
<span class="sd">            first_position (int): 1-indexed first position number.</span>
<span class="sd">            last_position (int): Optional last position to select a range or -1 to specify the end of the contig.</span>
<span class="sd">            filler_value (str): Optional filler for undefined regions beyond the genome data. Does not modify the data.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Returns the nucleotide at first_position, list of values from</span>
<span class="sd">            first_position to last_position inclusive, or None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">contig_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">set_current_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="n">queried_value</span> <span class="o">=</span> <span class="n">filler_value</span>
        <span class="k">if</span> <span class="n">last_position</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">first_position</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">]):</span>
                <span class="n">queried_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">][</span><span class="n">first_position</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">queried_value</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">if</span> <span class="n">last_position</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
                <span class="n">last_position</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">last_position</span> <span class="o">&gt;=</span> <span class="n">first_position</span> <span class="ow">and</span> <span class="n">first_position</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">]):</span>
                <span class="n">queried_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">][</span><span class="n">first_position</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span><span class="n">last_position</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">filler_value</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">queried_value</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">last_position</span> <span class="o">-</span> <span class="n">first_position</span> <span class="o">+</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">queried_value</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">filler_value</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span> <span class="n">last_position</span> <span class="o">-</span> <span class="n">first_position</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">queried_value</span><span class="p">)</span> <span class="p">))</span>
        <span class="k">return</span> <span class="n">queried_value</span>
</div>
<div class="viewcode-block" id="GenomeStatus.get_contig_length"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.get_contig_length">[docs]</a>    <span class="k">def</span> <span class="nf">get_contig_length</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            contig_name (str): Unique contig description.</span>

<span class="sd">        Returns:</span>
<span class="sd">            int: Number of positions defined in the contig</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">contig_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">set_current_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">contig_name</span><span class="p">])</span>
</div>
<div class="viewcode-block" id="GenomeStatus.send_to_fasta_handle"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.send_to_fasta_handle">[docs]</a>    <span class="k">def</span> <span class="nf">send_to_fasta_handle</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_handle</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">max_chars_per_line</span><span class="o">=</span><span class="mi">80</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Assumes the genome data is in string format or stringifiable and one</span>
<span class="sd">        character per position, and then writes it in to the handle open for</span>
<span class="sd">        writing.  The file format is like a typical fasta were the genome</span>
<span class="sd">        data to be base calls (but no checks are performed).</span>

<span class="sd">        The contigs are sorted by name, not the order they were created.</span>

<span class="sd">        Args:</span>
<span class="sd">            output_handle (file object): File to append FASTA string.</span>
<span class="sd">            contig_prefix (str): Prefix for all contig names.</span>
<span class="sd">            max_chars_per_line (int): A positive value will limit the max chars per line.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">current_contig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_contigs</span><span class="p">():</span>
            <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">contig_prefix</span> <span class="o">+</span> <span class="n">current_contig</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">max_chars_per_line</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="k">while</span> <span class="p">(</span> <span class="n">max_chars_per_line</span> <span class="o">*</span> <span class="n">i</span> <span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">current_contig</span><span class="p">]):</span>
                    <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">current_contig</span><span class="p">][</span>
                                                <span class="p">(</span> <span class="n">max_chars_per_line</span> <span class="o">*</span> <span class="n">i</span> <span class="p">):(</span> <span class="n">max_chars_per_line</span> <span class="o">*</span> <span class="p">(</span> <span class="n">i</span> <span class="o">+</span> <span class="mi">1</span> <span class="p">)</span> <span class="p">)])</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                    <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">output_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span><span class="p">[</span><span class="n">current_contig</span><span class="p">]</span> <span class="o">+</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">))</span>
</div>
<div class="viewcode-block" id="GenomeStatus.write_to_fasta_file"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeStatus.write_to_fasta_file">[docs]</a>    <span class="k">def</span> <span class="nf">write_to_fasta_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_filename</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">max_chars_per_line</span><span class="o">=</span><span class="mi">80</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Opens the passed filename and passes to send_to_fasta_handle.</span>
<span class="sd">        This is a separate function so that unit testing is easier, and</span>
<span class="sd">        file names or open file handles can be used as destinations.</span>

<span class="sd">        Args:</span>
<span class="sd">            output_filename (str): Output filename.</span>
<span class="sd">            contig_prefix (str): Prefix for all contig names.</span>
<span class="sd">            max_chars_per_line (int): A positive value will limit the max contig chars per line.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output_filename</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output_handle</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">send_to_fasta_handle</span><span class="p">(</span><span class="n">output_handle</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="p">,</span> <span class="n">max_chars_per_line</span><span class="p">)</span>

</div></div>
<div class="viewcode-block" id="Genome"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.Genome">[docs]</a><span class="k">class</span> <span class="nc">Genome</span><span class="p">(</span><span class="n">GenomeStatus</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A special type of GenomeStatus where the genome information being stored</span>
<span class="sd">    is always actual base calls, as strings.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attributes:</span>
<span class="sd">            _genome is an alias of _status_data, provided for code clarity</span>
<span class="sd">            when working with an actual genome</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">GenomeStatus</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genome</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_status_data</span>

<div class="viewcode-block" id="Genome.set_call"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.Genome.set_call">[docs]</a>    <span class="k">def</span> <span class="nf">set_call</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">new_data</span><span class="p">,</span> <span class="n">first_position</span><span class="p">,</span> <span class="n">missing_range_filler</span><span class="o">=</span><span class="s">&quot;X&quot;</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Alias of set_value, for code clarity &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_value</span><span class="p">(</span><span class="n">new_data</span><span class="p">,</span> <span class="n">first_position</span><span class="p">,</span> <span class="n">missing_range_filler</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="Genome.get_call"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.Genome.get_call">[docs]</a>    <span class="k">def</span> <span class="nf">get_call</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">first_position</span><span class="p">,</span> <span class="n">last_position</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">filler_value</span><span class="o">=</span><span class="s">&quot;X&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Alias of get_value, for code clarity &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span> <span class="n">first_position</span><span class="p">,</span> <span class="n">last_position</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">,</span> <span class="n">filler_value</span><span class="p">)</span>
</div>
    <span class="k">def</span> <span class="nf">_import_fasta_line</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line_from_fasta</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Assumes the string passed in is a line from a fasta file, and</span>
<span class="sd">        populates the genome with the information contained.  Not meant</span>
<span class="sd">        to be called on any data except a full fasta file in order by</span>
<span class="sd">        line top to bottom.</span>

<span class="sd">        Args:</span>
<span class="sd">            line_from_fasta (str): the current line to parse</span>
<span class="sd">            contig_prefix (str): the prefix will be removed from the parsed contig name</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">import</span> <span class="nn">re</span>

        <span class="c"># Parse the contig name discarding the prefix and surrounding whitespace characters</span>
        <span class="n">contig_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^&gt;&#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">(</span><span class="n">contig_prefix</span><span class="p">)</span> <span class="o">+</span> <span class="s">r&#39;([^\s]+)(?:\s|$)&#39;</span><span class="p">,</span> <span class="n">line_from_fasta</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">contig_match</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_contig</span><span class="p">(</span><span class="n">contig_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c"># Parse the contig sequence discarding trailing whitespace characters</span>
            <span class="n">data_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^([A-Za-z.-]+)\s*$&#39;</span><span class="p">,</span> <span class="n">line_from_fasta</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">data_match</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">append_contig</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">data_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)))</span>

    <span class="c"># TODO (jtravis): Remove the set_current_contig methods and make the contig_name parameters required.</span>
    <span class="c"># It looks like the set_current_contig method was created to remember the contig_name while parsing</span>
    <span class="c"># the fasta files. It doesn&#39;t make sense to be calling it everywhere instead of simply using the contig_name.</span>
    <span class="c"># Replace the set_current_contig method with a variable as in this alternative implementation of import_fasta_file</span>
    <span class="c"># and _import_fasta_line. This implementation ignores unused contig_prefix parameter and character restriction</span>
    <span class="c"># guards of the current implementation.</span>
    <span class="c">#</span>
    <span class="c"># def import_fasta_file(self, filename):</span>
    <span class="c">#   with open(filename, &#39;r&#39;) as handle:</span>
    <span class="c">#        self._import_fasta_file(handle)</span>
    <span class="c">#</span>
    <span class="c"># def _import_fasta_file(self, handle):</span>
    <span class="c">#   contig_name = &quot;&quot;</span>
    <span class="c">#   for line in handle:</span>
    <span class="c">#       line = line.strip()</span>
    <span class="c">#       if line.startswith(&#39;&gt;&#39;):</span>
    <span class="c">#           contig_name = line[1:]</span>
    <span class="c">#       else:</span>
    <span class="c">#           self.append_contig(contig_name, line)</span>
<div class="viewcode-block" id="Genome.import_fasta_file"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.Genome.import_fasta_file">[docs]</a>    <span class="k">def</span> <span class="nf">import_fasta_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fasta_filename</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Read in a fasta file.</span>

<span class="sd">        Args:</span>
<span class="sd">            fasta_filename (str): fasta file to import</span>
<span class="sd">            contig_prefix (str): the prefix will be removed from the parsed contig names</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fasta_filename</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fasta_handle</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line_from_fasta</span> <span class="ow">in</span> <span class="n">fasta_handle</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_import_fasta_line</span><span class="p">(</span><span class="n">line_from_fasta</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="p">)</span>
</div>
    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Genome.reverse_complement"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.Genome.reverse_complement">[docs]</a>    <span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="n">dna_string</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            dna_string (str): nucleotide sequence to reverse complement</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; Genome.reverse_complement(&#39;GATC&#39;)</span>
<span class="sd">            CTAG</span>

<span class="sd">        Returns:</span>
<span class="sd">            string: nucleotide sequence reverse complement</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">dna_string</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
            <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="s">&#39;ABCDGHMNRSTUVWXYabcdghmnrstuvwxy&#39;</span><span class="p">,</span> <span class="s">&#39;TVGHCDKNYSAABWXRtvghcdknysaabwxr&#39;</span><span class="p">))[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</div>
    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="Genome.simple_call"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.Genome.simple_call">[docs]</a>    <span class="k">def</span> <span class="nf">simple_call</span><span class="p">(</span><span class="n">dna_string</span><span class="p">,</span> <span class="n">allow_x</span><span class="o">=</span><span class="bp">False</span><span class="p">,</span> <span class="n">allow_del</span><span class="o">=</span><span class="bp">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Standardizes the DNA call assumed to be the base at position one.</span>
<span class="sd">        Discards insertion data, changes &#39;U&#39; to &#39;T&#39;, and changes degeneracies to &#39;N&#39;.</span>
<span class="sd">        &#39;X&#39; and deletes are changed to &#39;N&#39; by default.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; Genome.simple_call(&quot;GATC&quot;)</span>
<span class="sd">            G</span>
<span class="sd">            &gt;&gt;&gt; Genome.simple_call(&quot;X&quot;)</span>
<span class="sd">            N</span>
<span class="sd">            &gt;&gt;&gt; Genome.simple_call(&quot;U&quot;)</span>
<span class="sd">            T</span>

<span class="sd">        Args:</span>
<span class="sd">            dna_string (str): Only the first position is considered.</span>
<span class="sd">            allow_x (bool): False &#39;X&#39; are replaced with &#39;N&#39;.</span>
<span class="sd">            allow_del (bool): False &#39;.&#39; are replaced with &#39;N&#39;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            string: &#39;A&#39;, &#39;C&#39;, &#39;G&#39;, &#39;T&#39;, or &#39;N&#39; with optional &#39;X&#39; and &#39;.&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">simple_base</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">dna_string</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">simple_base</span> <span class="o">=</span> <span class="n">dna_string</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">allow_del</span><span class="p">:</span>
            <span class="n">simple_base</span> <span class="o">=</span> <span class="s">&#39;.&#39;</span>
        <span class="k">if</span> <span class="n">simple_base</span> <span class="o">==</span> <span class="s">&#39;U&#39;</span><span class="p">:</span>
            <span class="c"># RNA Uracil transcribes to DNA Thymine</span>
            <span class="n">simple_base</span> <span class="o">=</span> <span class="s">&#39;T&#39;</span>
        <span class="k">if</span> <span class="n">simple_base</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">,</span> <span class="s">&#39;.&#39;</span><span class="p">]:</span>
            <span class="c"># Change degeneracies to &#39;N&#39;</span>
            <span class="n">simple_base</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">allow_x</span> <span class="ow">and</span> <span class="p">(</span> <span class="n">simple_base</span> <span class="o">==</span> <span class="s">&#39;X&#39;</span> <span class="p">):</span>
            <span class="n">simple_base</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">allow_del</span> <span class="ow">and</span> <span class="p">(</span> <span class="n">simple_base</span> <span class="o">==</span> <span class="s">&#39;.&#39;</span> <span class="p">):</span>
            <span class="n">simple_base</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
        <span class="k">return</span> <span class="n">simple_base</span>

</div></div>
<div class="viewcode-block" id="GenomeMeta"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta">[docs]</a><span class="k">class</span> <span class="nc">GenomeMeta</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Stores the metadata associated with a genome.  &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attributes:</span>
<span class="sd">            _nickname: is our best guess at the name information as supplied by the</span>
<span class="sd">            input files.  Its determination depends on filetype.  For example, the</span>
<span class="sd">            nickname of a genome from a fasta file is the filename minus the</span>
<span class="sd">            extension.  There is no expectation that this value is unique.  A</span>
<span class="sd">            ( _file_path, _nickname ) tuple or the like would stand a much higher</span>
<span class="sd">            chance of being unique.</span>
<span class="sd">            _filepath:</span>
<span class="sd">            _file_type:</span>
<span class="sd">            _generators: A list of analysis tools that have been run on input files to produce</span>
<span class="sd">            this data, from earliest to latest.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_nickname</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_file_path</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_file_type</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_generators</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c"># TODO: This should probably be a dictionary someday</span>

    <span class="c"># NOTE(jtravis): side effect alters nickname</span>
<div class="viewcode-block" id="GenomeMeta.set_file_path"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.set_file_path">[docs]</a>    <span class="k">def</span> <span class="nf">set_file_path</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            file_path (str):</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_file_path</span> <span class="o">=</span> <span class="n">file_path</span>
        <span class="c"># TODO (jtravis): merge and replace if-statement with set_nickname()</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_nickname</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_nickname</span> <span class="o">=</span> <span class="n">GenomeMeta</span><span class="o">.</span><span class="n">generate_nickname_from_filename</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="GenomeMeta.set_file_type"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.set_file_type">[docs]</a>    <span class="k">def</span> <span class="nf">set_file_type</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">type_string</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            type_string (str): frankenfasta or vcf</span>

<span class="sd">        See Also:</span>
<span class="sd">            nasp.matrix_DTO._parse_files</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_file_type</span> <span class="o">=</span> <span class="n">type_string</span>
</div>
<div class="viewcode-block" id="GenomeMeta.set_nickname"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.set_nickname">[docs]</a>    <span class="k">def</span> <span class="nf">set_nickname</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">nickname</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This value should be unique per-file, for multi-sample input files,</span>
<span class="sd">        but there is no expectation that this would be unique per run, and</span>
<span class="sd">        nothing to enforce even per-file uniqueness.</span>

<span class="sd">        Args:</span>
<span class="sd">            nickname (str):</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_nickname</span> <span class="o">=</span> <span class="n">nickname</span>
</div>
<div class="viewcode-block" id="GenomeMeta.add_generators"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.add_generators">[docs]</a>    <span class="k">def</span> <span class="nf">add_generators</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">generator_array</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            generator_array: A list of analysis tools that have been run on input files to produce</span>
<span class="sd">            this data, from earliest to latest.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_generators</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">generator_array</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="GenomeMeta.file_path"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.file_path">[docs]</a>    <span class="k">def</span> <span class="nf">file_path</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns:</span>
<span class="sd">            string: Path to the source file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_file_path</span>
</div>
<div class="viewcode-block" id="GenomeMeta.file_type"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.file_type">[docs]</a>    <span class="k">def</span> <span class="nf">file_type</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns:</span>
<span class="sd">            string:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_file_type</span>
</div>
<div class="viewcode-block" id="GenomeMeta.nickname"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.nickname">[docs]</a>    <span class="k">def</span> <span class="nf">nickname</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Returns:</span>
<span class="sd">            string: Basename without the file extension</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_nickname</span>
</div>
<div class="viewcode-block" id="GenomeMeta.identifier"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.identifier">[docs]</a>    <span class="k">def</span> <span class="nf">identifier</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Format: nickname or nickname::generator,generator,...</span>

<span class="sd">        Example:</span>


<span class="sd">        Returns:</span>
<span class="sd">            string: meant to be recognizable to the user to differentiate</span>
<span class="sd">            their sample-analyses.  There can be no assumption that this value is</span>
<span class="sd">            unique, because the nickname is often not unique.  As a result, this</span>
<span class="sd">            should not be used in-code (EG dictionary keys) to differentiate</span>
<span class="sd">            samples.  The best we can do is a ( _file_path, _nickname ) tuple, and</span>
<span class="sd">            even that may not be guaranteed to be unique.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">identifier</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_nickname</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_generators</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">identifier</span> <span class="o">=</span> <span class="s">&quot;{0}::{1}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">identifier</span><span class="p">,</span> <span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_generators</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">identifier</span>
</div>
    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="GenomeMeta.generate_nickname_from_filename"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.generate_nickname_from_filename">[docs]</a>    <span class="k">def</span> <span class="nf">generate_nickname_from_filename</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        For single-sample input files that don&#39;t carry any sample name</span>
<span class="sd">        metadata within the file, generate a nickname for the sample by</span>
<span class="sd">        removing the extension.  If this fails, generate a random name in the</span>
<span class="sd">        format &quot;file_XXXXXXXX&quot; where X is an 8-digit random integer.</span>

<span class="sd">        Args:</span>
<span class="sd">            filename (str):</span>

<span class="sd">        Returns:</span>
<span class="sd">            string: filename sans extension or &quot;file_XXXXXXXX&quot; where X is an 8-digit random integer.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">import</span> <span class="nn">re</span>
        <span class="kn">import</span> <span class="nn">random</span>
        <span class="c"># Parse basename from fasta or vcf file</span>
        <span class="n">filename_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^(?:.*/)?([^/]+?)\.(?:(?:franken)?fas?(?:ta)?|vcf)?$&#39;</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">re</span><span class="o">.</span><span class="n">IGNORECASE</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">filename_match</span><span class="p">:</span>
            <span class="n">nickname</span> <span class="o">=</span> <span class="n">filename_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c"># postfix file with 8 digit random number where 0&#39;s are allowed</span>
            <span class="n">nickname</span> <span class="o">=</span> <span class="s">&quot;file_&quot;</span> <span class="o">+</span> <span class="s">&quot;</span><span class="si">%08d</span><span class="s">&quot;</span> <span class="o">%</span> <span class="n">random</span><span class="o">.</span><span class="n">randrange</span><span class="p">(</span><span class="mi">10</span> <span class="o">**</span> <span class="mi">8</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">nickname</span>
</div>
    <span class="nd">@staticmethod</span>
<div class="viewcode-block" id="GenomeMeta.reverse_complement"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeMeta.reverse_complement">[docs]</a>    <span class="k">def</span> <span class="nf">reverse_complement</span><span class="p">(</span><span class="n">dna_string</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">dna_string</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
            <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">maketrans</span><span class="p">(</span><span class="s">&#39;ABCDGHMNRSTUVWXYabcdghmnrstuvwxy&#39;</span><span class="p">,</span> <span class="s">&#39;TVGHCDKNYSAABWXRtvghcdknysaabwxr&#39;</span><span class="p">))[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>

</div></div>
<div class="viewcode-block" id="IndelList"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.IndelList">[docs]</a><span class="k">class</span> <span class="nc">IndelList</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    For storing indel data separately from the reference-indexed position</span>
<span class="sd">    data.  Mostly a relic from when calls were stored as a long string with</span>
<span class="sd">    one character per position.  Might still have some use for indel</span>
<span class="sd">    implementation, but might be no longer useful.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attributes:</span>
<span class="sd">            _indels (dict):</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_indels</span> <span class="o">=</span> <span class="p">{}</span>

</div>
<div class="viewcode-block" id="ReferenceGenome"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.ReferenceGenome">[docs]</a><span class="k">class</span> <span class="nc">ReferenceGenome</span><span class="p">(</span><span class="n">Genome</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A special type of genome that is to be used as our reference.</span>
<span class="sd">    Unlike other genomes, we know there will only be one, it carries duplicate</span>
<span class="sd">    region data, and we don&#39;t need to store any metadata about it.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Attributes:</span>
<span class="sd">            _dups (GenomeStatus): carries data about whether a particular</span>
<span class="sd">            region of the reference was found to be very similar to another region</span>
<span class="sd">            in the same reference.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">Genome</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_dups</span> <span class="o">=</span> <span class="n">GenomeStatus</span><span class="p">()</span>

    <span class="c"># NOTE: Unused parameter last_position. It is always None.</span>
<div class="viewcode-block" id="ReferenceGenome.get_dups_call"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.ReferenceGenome.get_dups_call">[docs]</a>    <span class="k">def</span> <span class="nf">get_dups_call</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">first_position</span><span class="p">,</span> <span class="n">last_position</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            first_position (int):</span>
<span class="sd">            last_position (int):</span>
<span class="sd">            contig_name (str):</span>

<span class="sd">        Returns:</span>
<span class="sd">            str:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_dups</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="n">first_position</span><span class="p">,</span> <span class="n">last_position</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">)</span>

    <span class="c"># NOTE: unused parameter contig_prefix</span></div>
    <span class="k">def</span> <span class="nf">_import_dups_line</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line_from_dups_file</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Just like importing any other fasta-like file line-by-line, but</span>
<span class="sd">        specific to duplicate region data.</span>

<span class="sd">        Args:</span>
<span class="sd">            line_from_dups_file (str):</span>
<span class="sd">            contig_prefix (str):</span>

<span class="sd">        See Also:</span>
<span class="sd">            nasp.find_duplicates</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">import</span> <span class="nn">re</span>
        <span class="c"># Parse the contig name discarding the prefix and surrounding whitespace characters</span>
        <span class="n">contig_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^&gt;&#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">escape</span><span class="p">(</span><span class="n">contig_prefix</span><span class="p">)</span> <span class="o">+</span> <span class="s">r&#39;([^\s]+)(?:\s|$)&#39;</span><span class="p">,</span> <span class="n">line_from_dups_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">contig_match</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">add_contig</span><span class="p">(</span><span class="n">contig_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_dups</span><span class="o">.</span><span class="n">add_contig</span><span class="p">(</span><span class="n">contig_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c"># Parse the duplicate region sequence discarding trailing whitespace characters</span>
            <span class="n">data_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^([01-]+)\s*$&#39;</span><span class="p">,</span> <span class="n">line_from_dups_file</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">data_match</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_dups</span><span class="o">.</span><span class="n">append_contig</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">data_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)))</span>

    <span class="c"># NOTE: unused parameter contig_prefix</span>
<div class="viewcode-block" id="ReferenceGenome.import_dups_file"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.ReferenceGenome.import_dups_file">[docs]</a>    <span class="k">def</span> <span class="nf">import_dups_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dups_filename</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Wrapper for _import_dups_line for flexibility and testing.</span>

<span class="sd">        Args:</span>
<span class="sd">            dups_filename (str): Path to a fasta-style file where each contig position is either a 1 or 0 representing</span>
<span class="sd">             the presence or absence of a duplicate region.</span>
<span class="sd">            contig_prefix (str):</span>

<span class="sd">        See Also:</span>
<span class="sd">            nasp.find_duplicates</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">dups_filename</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">dups_handle</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line_from_dups_file</span> <span class="ow">in</span> <span class="n">dups_handle</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_import_dups_line</span><span class="p">(</span><span class="n">line_from_dups_file</span><span class="p">,</span> <span class="n">contig_prefix</span><span class="p">)</span>
</div></div>
<div class="viewcode-block" id="FastaGenome"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.FastaGenome">[docs]</a><span class="k">class</span> <span class="nc">FastaGenome</span><span class="p">(</span><span class="n">Genome</span><span class="p">,</span> <span class="n">GenomeMeta</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A special type of genome where we know the data came from a fasta file,</span>
<span class="sd">    and so we can omit the depth and proportion filters.  Meant to mimic</span>
<span class="sd">    a VCFGenome object, with filter checks hard-coded.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">Genome</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">GenomeMeta</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_indels</span> <span class="o">=</span> <span class="n">IndelList</span><span class="p">()</span>

    <span class="c"># FIXME This data should be put into a real structure when the fasta code is pulled in</span>
<div class="viewcode-block" id="FastaGenome.get_was_called"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.FastaGenome.get_was_called">[docs]</a>    <span class="k">def</span> <span class="nf">get_was_called</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        A stand-in for the was-called filter that just makes sure the position</span>
<span class="sd">        isn&#39;t an &quot;N&quot;.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: &#39;N&#39; if the call is an X or N, otherwise &#39;Y&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="s">&quot;N&quot;</span>
        <span class="n">call_to_check</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_call</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">,</span> <span class="s">&quot;X&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">call_to_check</span> <span class="o">!=</span> <span class="s">&quot;X&quot;</span> <span class="ow">and</span> <span class="n">call_to_check</span> <span class="o">!=</span> <span class="s">&quot;N&quot;</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">=</span> <span class="s">&quot;Y&quot;</span>
        <span class="k">return</span> <span class="n">return_value</span>
</div>
<div class="viewcode-block" id="FastaGenome.get_coverage_pass"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.FastaGenome.get_coverage_pass">[docs]</a>    <span class="k">def</span> <span class="nf">get_coverage_pass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; This filter is not applicable.</span>

<span class="sd">        Args:</span>
<span class="sd">            current_pos (int):</span>
<span class="sd">            contig_name (str):</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: &#39;-&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s">&quot;-&quot;</span>
</div>
<div class="viewcode-block" id="FastaGenome.get_proportion_pass"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.FastaGenome.get_proportion_pass">[docs]</a>    <span class="k">def</span> <span class="nf">get_proportion_pass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; This filter is not applicable.</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: &#39;-&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s">&quot;-&quot;</span>

</div></div>
<div class="viewcode-block" id="VCFGenome"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFGenome">[docs]</a><span class="k">class</span> <span class="nc">VCFGenome</span><span class="p">(</span><span class="n">Genome</span><span class="p">,</span> <span class="n">GenomeMeta</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A standard sample for analysis.  Has genome data, metadata, and data for</span>
<span class="sd">    the three filters.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">Genome</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">GenomeMeta</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_indels</span> <span class="o">=</span> <span class="n">IndelList</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_was_called</span> <span class="o">=</span> <span class="n">GenomeStatus</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_passed_coverage</span> <span class="o">=</span> <span class="n">GenomeStatus</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_passed_proportion</span> <span class="o">=</span> <span class="n">GenomeStatus</span><span class="p">()</span>

<div class="viewcode-block" id="VCFGenome.set_was_called"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFGenome.set_was_called">[docs]</a>    <span class="k">def</span> <span class="nf">set_was_called</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pass_value</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_was_called</span><span class="o">.</span><span class="n">set_value</span><span class="p">(</span><span class="n">pass_value</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="s">&quot;N&quot;</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="VCFGenome.set_coverage_pass"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFGenome.set_coverage_pass">[docs]</a>    <span class="k">def</span> <span class="nf">set_coverage_pass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pass_value</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_passed_coverage</span><span class="o">.</span><span class="n">set_value</span><span class="p">(</span><span class="n">pass_value</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="VCFGenome.set_proportion_pass"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFGenome.set_proportion_pass">[docs]</a>    <span class="k">def</span> <span class="nf">set_proportion_pass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pass_value</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_passed_proportion</span><span class="o">.</span><span class="n">set_value</span><span class="p">(</span><span class="n">pass_value</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="VCFGenome.get_was_called"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFGenome.get_was_called">[docs]</a>    <span class="k">def</span> <span class="nf">get_was_called</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_was_called</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">,</span> <span class="s">&quot;N&quot;</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="VCFGenome.get_coverage_pass"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFGenome.get_coverage_pass">[docs]</a>    <span class="k">def</span> <span class="nf">get_coverage_pass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_passed_coverage</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="VCFGenome.get_proportion_pass"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFGenome.get_proportion_pass">[docs]</a>    <span class="k">def</span> <span class="nf">get_proportion_pass</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_passed_proportion</span><span class="o">.</span><span class="n">get_value</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">,</span> <span class="s">&quot;?&quot;</span><span class="p">)</span>

</div></div>
<div class="viewcode-block" id="CollectionStatistics"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.CollectionStatistics">[docs]</a><span class="k">class</span> <span class="nc">CollectionStatistics</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Stores a running tally for the statistics for the run.</span>
<span class="sd">    Stats are in two categories: per-contig and per-sample.</span>
<span class="sd">    Contig stats are basic counts, and percentages based on reference length.</span>
<span class="sd">    Sample stats are tallied as x out of y for each position, and then</span>
<span class="sd">    counts for each sample, all samples, and any samples, can be computed</span>
<span class="sd">    automatically.</span>
<span class="sd">    For this reason, the object needs to know when the run moves on to the</span>
<span class="sd">    next position, and the flush_cumulative_stat_cache function does this.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        _cumulative_cache keeps a running tally of sample stats at the</span>
<span class="sd">        current position, and then writes the results to _sample_stats when</span>
<span class="sd">        flush_cumulative_stat_cache() is called.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig_stats</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">def</span> <span class="nf">_increment_by_contig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Makes sure the contig stat is defined, then increment it. &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span> <span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig_stats</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_contig_stats</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span> <span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig_stats</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span> <span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>

<div class="viewcode-block" id="CollectionStatistics.increment_contig_stat"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.CollectionStatistics.increment_contig_stat">[docs]</a>    <span class="k">def</span> <span class="nf">increment_contig_stat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Increments the stat for the specified contig, and automatically</span>
<span class="sd">        increments the count for the all-contigs tally on the same stat.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_increment_by_contig</span><span class="p">(</span><span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">contig_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_increment_by_contig</span><span class="p">(</span><span class="n">stat_id</span><span class="p">,</span> <span class="bp">None</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="CollectionStatistics.get_contig_stat"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.CollectionStatistics.get_contig_stat">[docs]</a>    <span class="k">def</span> <span class="nf">get_contig_stat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span> <span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig_stats</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig_stats</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">contig_name</span> <span class="p">)]</span>
        <span class="k">return</span> <span class="n">return_value</span>
</div>
    <span class="k">def</span> <span class="nf">_cache_cumulative_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">did_pass</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes data to the stat cache for the current position.  stat_type</span>
<span class="sd">        is either &quot;p&quot; for pass/positive or &quot;t&quot; for total.  Flushing the cache</span>
<span class="sd">        involves checking the p/t ratio.  When updating the cache,</span>
<span class="sd">        failed/negative samples just increment t, while pass/positive samples</span>
<span class="sd">        increment p and t.</span>
<span class="sd">        Cumulative stats are across all samples at a position, and then once</span>
<span class="sd">        each for all sample-analyses on a sample.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;t&#39;</span> <span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;t&#39;</span> <span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;p&#39;</span> <span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;t&#39;</span> <span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">did_pass</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;p&#39;</span> <span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="k">def</span> <span class="nf">_increment_by_sample</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">sample_info</span><span class="p">,</span> <span class="n">cum_type</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Defines if necessary, then increments, the sample stat. &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">sample_info</span><span class="p">,</span> <span class="n">cum_type</span> <span class="p">)</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">sample_info</span><span class="p">,</span> <span class="n">cum_type</span> <span class="p">)]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">sample_info</span><span class="p">,</span> <span class="n">cum_type</span> <span class="p">)]</span> <span class="o">+=</span> <span class="mi">1</span>

<div class="viewcode-block" id="CollectionStatistics.record_sample_stat"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.CollectionStatistics.record_sample_stat">[docs]</a>    <span class="k">def</span> <span class="nf">record_sample_stat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span><span class="p">,</span> <span class="n">did_pass</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Updates the sample stat, and then the cumulative stat cache.</span>
<span class="sd">        Increments the cache for all samples, and for all analyses</span>
<span class="sd">        on this sample.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">did_pass</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_increment_by_sample</span><span class="p">(</span><span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="p">(</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span> <span class="p">),</span> <span class="bp">None</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cache_cumulative_stats</span><span class="p">(</span><span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">did_pass</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cache_cumulative_stats</span><span class="p">(</span><span class="n">stat_id</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">did_pass</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="CollectionStatistics.get_sample_stat"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.CollectionStatistics.get_sample_stat">[docs]</a>    <span class="k">def</span> <span class="nf">get_sample_stat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span><span class="p">):</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="p">(</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span> <span class="p">),</span> <span class="bp">None</span> <span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="p">(</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span> <span class="p">),</span> <span class="bp">None</span> <span class="p">)]</span>
        <span class="k">return</span> <span class="n">return_value</span>
</div>
<div class="viewcode-block" id="CollectionStatistics.get_cumulative_stat"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.CollectionStatistics.get_cumulative_stat">[docs]</a>    <span class="k">def</span> <span class="nf">get_cumulative_stat</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">cum_type</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            stat_id (str):</span>
<span class="sd">            cum_type (str): &#39;any&#39; or &#39;all&#39;</span>
<span class="sd">            sample_nickname (str):</span>

<span class="sd">        Return:</span>
<span class="sd">            int:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">cum_type</span> <span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_stats</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">cum_type</span> <span class="p">)]</span>
        <span class="k">return</span> <span class="n">return_value</span>
</div>
<div class="viewcode-block" id="CollectionStatistics.flush_cumulative_stat_cache"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.CollectionStatistics.flush_cumulative_stat_cache">[docs]</a>    <span class="k">def</span> <span class="nf">flush_cumulative_stat_cache</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Does the math on the pass/total ratio for the stat cache for the current</span>
<span class="sd">        position, writes that to the sample stats for the any/all counts, and</span>
<span class="sd">        clears the cache.</span>
<span class="sd">        p &gt; 0:  any++</span>
<span class="sd">        p = t:  all++</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="n">stat_type</span> <span class="p">)</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">stat_type</span> <span class="o">==</span> <span class="s">&#39;p&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;p&#39;</span> <span class="p">)]</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">[</span>
                    <span class="p">(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;t&#39;</span> <span class="p">)]:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_increment_by_sample</span><span class="p">(</span><span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="s">&#39;all&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span><span class="p">[(</span> <span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="s">&#39;p&#39;</span> <span class="p">)]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_increment_by_sample</span><span class="p">(</span><span class="n">stat_id</span><span class="p">,</span> <span class="n">sample_nickname</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="s">&#39;any&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cumulative_cache</span> <span class="o">=</span> <span class="p">{}</span>

</div></div>
<div class="viewcode-block" id="GenomeCollection"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection">[docs]</a><span class="k">class</span> <span class="nc">GenomeCollection</span><span class="p">(</span><span class="n">CollectionStatistics</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A &quot;master matrix&quot; object, of sorts.</span>
<span class="sd">    Carries all the data necessary to make a matrix, although some of it</span>
<span class="sd">    is computed on-the-fly as the matrix is actually written, for</span>
<span class="sd">    performance reasons.  Stats aren&#39;t available until after the matrix</span>
<span class="sd">    is written, for this reason.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        _failed_genomes is just a list of filenames that had errors during</span>
<span class="sd">        import.  Added as blank columns to inform the user.</span>

<span class="sd">        Attributes:</span>
<span class="sd">            _reference (ReferenceGenome)</span>
<span class="sd">            _genomes (list of Genome):</span>
<span class="sd">            _genome_identifiers:</span>
<span class="sd">            _failed_genomes (list): Genome filenames that could not be parsed.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">CollectionStatistics</span><span class="o">.</span><span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genome_identifiers</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_failed_genomes</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_get_key</span><span class="p">(</span><span class="n">genome</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        A wrapper getter for a genome&#39;s identifier, for the purpose of passing</span>
<span class="sd">        informative data to the sort function.  This is used to make sample</span>
<span class="sd">        order predictable and preserved between different runs.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">genome</span><span class="o">.</span><span class="n">identifier</span><span class="p">()</span>

<div class="viewcode-block" id="GenomeCollection.set_reference"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.set_reference">[docs]</a>    <span class="k">def</span> <span class="nf">set_reference</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span> <span class="o">=</span> <span class="n">reference</span>
</div>
<div class="viewcode-block" id="GenomeCollection.reference"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.reference">[docs]</a>    <span class="k">def</span> <span class="nf">reference</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span>

    <span class="c"># NOTE (jtravis): Unused method</span></div>
<div class="viewcode-block" id="GenomeCollection.get_dups_call"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.get_dups_call">[docs]</a>    <span class="k">def</span> <span class="nf">get_dups_call</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">first_position</span><span class="p">,</span> <span class="n">last_position</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">get_dups_call</span><span class="p">(</span><span class="n">first_position</span><span class="p">,</span> <span class="n">last_position</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="GenomeCollection.add_genome"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.add_genome">[docs]</a>    <span class="k">def</span> <span class="nf">add_genome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adds the genome to the collection, then makes sure the genome list is</span>
<span class="sd">        properly set and in order.</span>

<span class="sd">        Args:</span>
<span class="sd">            genome (Genome):</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
        <span class="n">genome_nickname</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">nickname</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">genome_nickname</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genome_identifiers</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_genome_identifiers</span><span class="p">[</span><span class="n">genome_nickname</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genome_identifiers</span><span class="p">[</span><span class="n">genome_nickname</span><span class="p">][(</span> <span class="n">genome</span><span class="o">.</span><span class="n">identifier</span><span class="p">(),</span> <span class="n">genome</span><span class="o">.</span><span class="n">file_path</span><span class="p">()</span> <span class="p">)]</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="n">GenomeCollection</span><span class="o">.</span><span class="n">_get_key</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="GenomeCollection.add_failed_genome"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.add_failed_genome">[docs]</a>    <span class="k">def</span> <span class="nf">add_failed_genome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Args:</span>
<span class="sd">            genome_path (str): Path to genome file that raised an exception during parsing.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">genome_path</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_failed_genomes</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_failed_genomes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">genome_path</span><span class="p">)</span>

    <span class="c"># It looks like this method was created to</span></div>
<div class="viewcode-block" id="GenomeCollection.set_current_contig"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.set_current_contig">[docs]</a>    <span class="k">def</span> <span class="nf">set_current_contig</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">contig_name</span><span class="p">):</span>
        <span class="n">contig_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">set_current_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="p">:</span>
            <span class="n">genome</span><span class="o">.</span><span class="n">set_current_contig</span><span class="p">(</span><span class="n">contig_name</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">contig_name</span>

    <span class="c"># TODO (jtravis): Remove method. The information can also be accessed by calling self.reference().get_contigs()</span>
    <span class="c"># Using the @property decorator, this could be refactored as GenomeCollection.reference.contigs</span></div>
<div class="viewcode-block" id="GenomeCollection.get_contigs"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.get_contigs">[docs]</a>    <span class="k">def</span> <span class="nf">get_contigs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">get_contigs</span><span class="p">()</span>

    <span class="c"># FIXME this function is starting to become a bit of a cluster</span></div>
    <span class="k">def</span> <span class="nf">_format_matrix_line</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">matrix_formats</span><span class="p">,</span> <span class="n">pattern_data</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        A matrix line represents all the sample data at one reference</span>
<span class="sd">        contig-position.</span>
<span class="sd">        This function contains a large portion of the ultimate NASP logic that</span>
<span class="sd">        goes into generating a matrix, like filter application and consensus</span>
<span class="sd">        checking.</span>
<span class="sd">        A section below must be executed linearly and single-threaded for all</span>
<span class="sd">        contig-position-sample-analyses.  This is the &quot;expensive loop&quot;.</span>
<span class="sd">        The monolithic nature of this function, its massive size, and the fact</span>
<span class="sd">        that it makes little distinction between calculating something and</span>
<span class="sd">        writing something to a file, are all problems that will probably need</span>
<span class="sd">        to eventually be corrected.  I fear, however, that there is a</span>
<span class="sd">        significant tradeoff with it works / it&#39;s fast to run / it was fast to</span>
<span class="sd">        develop / it&#39;s easily maintained / it&#39;s beautiful.</span>

<span class="sd">        Args:</span>
<span class="sd">            current_contig (str):</span>
<span class="sd">            current_pos (int):</span>
<span class="sd">            matrix_formats (dict):</span>
<span class="sd">            pattern_data (dict):</span>

<span class="sd">        Example:</span>
<span class="sd">            &gt;&gt;&gt; _format_matrix_line(&quot;NC_007790.1&quot;, 1, )</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c"># &lt;output_folder&gt;/matrices/master_matrix.tsv</span>
        <span class="c"># &lt;output_folder&gt;/matrices/bestsnp_matrix.tsv</span>
        <span class="c"># &lt;output_folder&gt;/matrices/missingdata_matrix.tsv</span>
        <span class="c"># &lt;output_folder&gt;/matrices/withallrefpos_matrix.tsv (optional)</span>
        <span class="n">all_matrices</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">all_vcfs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">allcallable_matrices</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">snp_matrices_best</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">snp_matrices_md</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">fasta_pending_data</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">vcf_pending_data</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">current_pattern</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
        <span class="c"># Key None stores next unused, value 1 reserved for reference</span>
        <span class="n">current_pattern_legend</span> <span class="o">=</span> <span class="p">{</span><span class="bp">None</span><span class="p">:</span> <span class="mi">2</span><span class="p">}</span>
        <span class="n">encountered_calls</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="c"># Initialize matrix arrays</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;matrix&#39;</span><span class="p">:</span>
                <span class="n">all_matrices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;allcallable&#39;</span><span class="p">:</span>
                    <span class="n">allcallable_matrices</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;bestsnp&#39;</span> <span class="ow">or</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;includeref&#39;</span><span class="p">:</span>
                    <span class="n">snp_matrices_best</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;missingdata&#39;</span><span class="p">:</span>
                    <span class="n">snp_matrices_md</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;vcf&#39;</span><span class="p">:</span>
                <span class="n">all_vcfs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">)</span>
        <span class="n">genome_count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="p">)</span>
        <span class="n">failed_genome_count</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_failed_genomes</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">all_matrices</span><span class="p">:</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;{0}::{1}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_contig</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">current_pos</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">all_vcfs</span><span class="p">:</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">.</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_contig</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">current_pos</span><span class="p">))</span>
        <span class="n">reference_call</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">get_call</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="n">simplified_refcall</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">simple_call</span><span class="p">(</span><span class="n">reference_call</span><span class="p">)</span>
        <span class="n">fasta_pending_data</span><span class="p">[</span><span class="s">&#39;(Reference)&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">simplified_refcall</span>
        <span class="k">if</span> <span class="n">simplified_refcall</span> <span class="o">==</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
            <span class="n">current_pattern</span> <span class="o">+=</span> <span class="s">&#39;N&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">current_pattern_legend</span><span class="p">[</span><span class="n">simplified_refcall</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">current_pattern</span> <span class="o">+=</span> <span class="s">&#39;1&#39;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;reference_length&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">simplified_refcall</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;reference_clean&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="p">(</span> <span class="n">all_matrices</span> <span class="o">+</span> <span class="n">all_vcfs</span> <span class="p">):</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">reference_call</span><span class="p">)</span>
        <span class="n">dups_call</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">get_dups_call</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">dups_call</span> <span class="o">==</span> <span class="s">&quot;1&quot;</span><span class="p">:</span>
            <span class="n">dups_call</span> <span class="o">=</span> <span class="bp">True</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;reference_duplicated&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">dups_call</span> <span class="o">=</span> <span class="bp">False</span>

        <span class="n">call_data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="c"># Tally of sample base calls</span>
            <span class="s">&#39;A&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;N&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="c"># Tally of indels</span>
            <span class="c"># NOTE: Unused key</span>
            <span class="s">&#39;indel&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="c"># Tally of positions that were one of A,C,G,T</span>
            <span class="s">&#39;snpcall&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="c"># NOTE: Unused key</span>
            <span class="s">&#39;indelcall&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="c"># Tally of sample positions that match the reference</span>
            <span class="s">&#39;refcall&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="c">#</span>
            <span class="s">&#39;callstring&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span><span class="p">,</span> <span class="s">&#39;covstring&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span><span class="p">,</span> <span class="s">&#39;propstring&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span><span class="p">,</span>
            <span class="c">#</span>
            <span class="s">&#39;called&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;passcov&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;passprop&#39;</span><span class="p">:</span> <span class="mi">0</span>
        <span class="p">}</span>
        <span class="n">consensus_check</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="c"># The expensive loop, single threaded and runs for every sample-analysis-contig-position</span>
        <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="p">:</span>
            <span class="n">sample_call</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">get_call</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">)</span>
            <span class="n">simplified_sample_call</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">simple_call</span><span class="p">(</span><span class="n">sample_call</span><span class="p">)</span>
            <span class="n">call_data</span><span class="p">[</span><span class="n">simplified_sample_call</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">genome_nickname</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">nickname</span><span class="p">()</span>
            <span class="n">genome_identifier</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">identifier</span><span class="p">()</span>
            <span class="n">genome_path</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">file_path</span><span class="p">()</span>
            <span class="n">was_called</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">get_was_called</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
            <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;callstring&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">was_called</span>
            <span class="k">if</span> <span class="n">was_called</span> <span class="o">==</span> <span class="s">&#39;Y&#39;</span><span class="p">:</span>
                <span class="n">was_called</span> <span class="o">=</span> <span class="bp">True</span>
                <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;called&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">was_called</span> <span class="o">=</span> <span class="bp">False</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;was_called&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="n">was_called</span><span class="p">)</span>
            <span class="n">passed_coverage</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">get_coverage_pass</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
            <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;covstring&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">passed_coverage</span>
            <span class="k">if</span> <span class="n">passed_coverage</span> <span class="o">==</span> <span class="s">&#39;Y&#39;</span> <span class="ow">or</span> <span class="n">passed_coverage</span> <span class="o">==</span> <span class="s">&#39;-&#39;</span><span class="p">:</span>
                <span class="n">passed_coverage</span> <span class="o">=</span> <span class="bp">True</span>
                <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;passcov&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">passed_coverage</span> <span class="o">=</span> <span class="bp">False</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;passed_coverage_filter&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span>
                                    <span class="n">passed_coverage</span><span class="p">)</span>
            <span class="n">passed_proportion</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">get_proportion_pass</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
            <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;propstring&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="n">passed_proportion</span>
            <span class="k">if</span> <span class="n">passed_proportion</span> <span class="o">==</span> <span class="s">&#39;Y&#39;</span> <span class="ow">or</span> <span class="n">passed_proportion</span> <span class="o">==</span> <span class="s">&#39;-&#39;</span><span class="p">:</span>
                <span class="n">passed_proportion</span> <span class="o">=</span> <span class="bp">True</span>
                <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;passprop&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">passed_proportion</span> <span class="o">=</span> <span class="bp">False</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;passed_proportion_filter&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span>
                                    <span class="n">passed_proportion</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">was_called</span> <span class="ow">and</span> <span class="n">passed_coverage</span> <span class="ow">and</span> <span class="n">passed_proportion</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">genome_nickname</span> <span class="ow">in</span> <span class="n">consensus_check</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">consensus_check</span><span class="p">[</span><span class="n">genome_nickname</span><span class="p">]</span> <span class="o">!=</span> <span class="n">simplified_sample_call</span><span class="p">:</span>
                        <span class="n">consensus_check</span><span class="p">[</span><span class="n">genome_nickname</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">consensus_check</span><span class="p">[</span><span class="n">genome_nickname</span><span class="p">]</span> <span class="o">=</span> <span class="n">simplified_sample_call</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">consensus_check</span><span class="p">[</span><span class="n">genome_nickname</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
            <span class="c"># FIXME indels</span>
            <span class="k">if</span> <span class="n">was_called</span> <span class="ow">and</span> <span class="n">passed_coverage</span> <span class="ow">and</span> <span class="n">passed_proportion</span> <span class="ow">and</span> <span class="n">simplified_refcall</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">dups_call</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;quality_breadth&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">simplified_refcall</span> <span class="o">==</span> <span class="n">simplified_sample_call</span><span class="p">:</span>
                    <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;refcall&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">dups_call</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_reference&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span>
                                                <span class="bp">True</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_snp&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_indel&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_degen&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">simplified_sample_call</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
                    <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">dups_call</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_reference&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span>
                                                <span class="bp">False</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_snp&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_indel&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_degen&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">dups_call</span><span class="p">:</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_reference&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span>
                                                <span class="bp">False</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_snp&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_indel&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
                        <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;called_degen&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">simplified_refcall</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">dups_call</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;quality_breadth&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="n">genome_identifier</span><span class="p">,</span> <span class="n">genome_path</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="p">(</span> <span class="n">allcallable_matrices</span> <span class="o">+</span> <span class="n">snp_matrices_best</span> <span class="p">):</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample_call</span><span class="p">)</span>
            <span class="n">fasta_pending_data</span><span class="p">[</span><span class="n">genome_identifier</span><span class="p">]</span> <span class="o">=</span> <span class="n">simplified_sample_call</span>
            <span class="n">vcf_current_data</span> <span class="o">=</span> <span class="p">{</span> <span class="s">&#39;GT&#39;</span><span class="p">:</span> <span class="s">&#39;.&#39;</span><span class="p">,</span> <span class="s">&#39;was_called&#39;</span><span class="p">:</span> <span class="n">was_called</span><span class="p">,</span> <span class="s">&#39;passed_coverage&#39;</span><span class="p">:</span> <span class="n">passed_coverage</span><span class="p">,</span> <span class="s">&#39;passed_proportion&#39;</span><span class="p">:</span> <span class="n">passed_proportion</span> <span class="p">}</span>
            <span class="k">if</span> <span class="n">was_called</span> <span class="ow">and</span> <span class="n">passed_coverage</span> <span class="ow">and</span> <span class="n">passed_proportion</span> <span class="ow">and</span> <span class="n">simplified_sample_call</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">simplified_sample_call</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">current_pattern_legend</span><span class="p">:</span>
                    <span class="n">current_pattern_legend</span><span class="p">[</span><span class="n">simplified_sample_call</span><span class="p">]</span> <span class="o">=</span> <span class="n">current_pattern_legend</span><span class="p">[</span><span class="bp">None</span><span class="p">]</span>
                    <span class="n">current_pattern_legend</span><span class="p">[</span><span class="bp">None</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="c"># FIXME indels</span>
                    <span class="n">encountered_calls</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simplified_sample_call</span><span class="p">)</span>
                <span class="n">current_pattern</span> <span class="o">+=</span> <span class="nb">str</span><span class="p">(</span><span class="n">current_pattern_legend</span><span class="p">[</span><span class="n">simplified_sample_call</span><span class="p">])</span>
                <span class="n">vcf_current_data</span><span class="p">[</span><span class="s">&#39;GT&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">current_pattern_legend</span><span class="p">[</span><span class="n">simplified_sample_call</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">snp_matrices_md</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample_call</span><span class="p">)</span>
            <span class="k">elif</span> <span class="ow">not</span> <span class="n">was_called</span><span class="p">:</span>
                <span class="n">current_pattern</span> <span class="o">+=</span> <span class="s">&#39;N&#39;</span>
                <span class="n">fasta_pending_data</span><span class="p">[</span><span class="n">genome_identifier</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
                <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">snp_matrices_md</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;X</span><span class="se">\t</span><span class="s">&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">current_pattern</span> <span class="o">+=</span> <span class="s">&#39;N&#39;</span>
                <span class="n">fasta_pending_data</span><span class="p">[</span><span class="n">genome_identifier</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;N&#39;</span>
                <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">snp_matrices_md</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;N</span><span class="se">\t</span><span class="s">&quot;</span>
            <span class="n">vcf_pending_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">vcf_current_data</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">all_matrices</span><span class="p">:</span>
            <span class="c"># If an error occurred reading a genome file, include a blank column with the filename as the header</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span> <span class="o">*</span> <span class="n">failed_genome_count</span>
        <span class="k">for</span> <span class="n">genome_nickname</span> <span class="ow">in</span> <span class="n">consensus_check</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">consensus_check</span><span class="p">[</span><span class="n">genome_nickname</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;consensus&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="bp">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">record_sample_stat</span><span class="p">(</span><span class="s">&#39;consensus&#39;</span><span class="p">,</span> <span class="n">genome_nickname</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="bp">False</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">all_matrices</span><span class="p">:</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">{2}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]),</span>
                                                                     <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;indelcall&#39;</span><span class="p">]),</span>
                                                                     <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;refcall&#39;</span><span class="p">]))</span>
            <span class="c"># #CallWasMade, #PassedDepthFilter, and #PassedProportionFilter tally columns</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}/{3}</span><span class="se">\t</span><span class="s">{1}/{3}</span><span class="se">\t</span><span class="s">{2}/{3}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;called&#39;</span><span class="p">]),</span>
                                                                                 <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;passcov&#39;</span><span class="p">]),</span>
                                                                                 <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;passprop&#39;</span><span class="p">]),</span>
                                                                                 <span class="nb">str</span><span class="p">(</span><span class="n">genome_count</span><span class="p">))</span>
            <span class="c"># #A, #C, #G, #T, #Indel, and #NXdegen tally columns</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">{2}</span><span class="se">\t</span><span class="s">{3}</span><span class="se">\t</span><span class="s">{4}</span><span class="se">\t</span><span class="s">{5}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">]),</span>
                                                                                    <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;C&#39;</span><span class="p">]),</span>
                                                                                    <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;G&#39;</span><span class="p">]),</span>
                                                                                    <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;T&#39;</span><span class="p">]),</span>
                                                                                    <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;indel&#39;</span><span class="p">]),</span>
                                                                                    <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;N&#39;</span><span class="p">]))</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_contig</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">current_pos</span><span class="p">))</span>
        <span class="k">if</span> <span class="s">&#39;N&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">consensus_check</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="n">consensus_check</span> <span class="o">=</span> <span class="bp">True</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;all_passed_consensus&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">consensus_check</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">if</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;called&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">genome_count</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;all_called&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;passcov&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">genome_count</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;all_passed_coverage&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;passprop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">genome_count</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;all_passed_proportion&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">consensus_check</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">dups_call</span> <span class="ow">and</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;called&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">genome_count</span> <span class="ow">and</span> <span class="n">call_data</span><span class="p">[</span>
            <span class="s">&#39;passcov&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">genome_count</span> <span class="ow">and</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;passprop&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">genome_count</span> <span class="ow">and</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;N&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;quality_breadth&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;best_snps&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">dups_call</span> <span class="ow">and</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">increment_contig_stat</span><span class="p">(</span><span class="s">&#39;any_snps&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">all_matrices</span><span class="p">:</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">dups_call</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">consensus_check</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="p">(</span> <span class="n">allcallable_matrices</span> <span class="o">+</span> <span class="n">snp_matrices_md</span> <span class="p">):</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">{2}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;callstring&#39;</span><span class="p">]),</span> <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;covstring&#39;</span><span class="p">]),</span> <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;propstring&#39;</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">all_matrices</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">current_pattern</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pattern_data</span><span class="p">:</span>
                <span class="n">pattern_data</span><span class="p">[</span><span class="n">current_pattern</span><span class="p">]</span> <span class="o">=</span> <span class="n">pattern_data</span><span class="p">[</span><span class="bp">None</span><span class="p">]</span>
                <span class="n">pattern_data</span><span class="p">[</span><span class="bp">None</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;&#39;{0}&#39;</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_pattern</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pattern_data</span><span class="p">[</span><span class="n">current_pattern</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">all_vcfs</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">encountered_calls</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">encountered_calls</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;.&quot;</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">.</span><span class="se">\t</span><span class="s">PASS</span><span class="se">\t</span><span class="s">AN={0};NS={1}</span><span class="se">\t</span><span class="s">GT:FT&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">encountered_calls</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;indelcall&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">call_data</span><span class="p">[</span><span class="s">&#39;refcall&#39;</span><span class="p">]))</span>
            <span class="k">for</span> <span class="n">vcf_current_data</span> <span class="ow">in</span> <span class="n">vcf_pending_data</span><span class="p">:</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;</span><span class="se">\t</span><span class="s">{0}:&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">vcf_current_data</span><span class="p">[</span><span class="s">&#39;GT&#39;</span><span class="p">]))</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">vcf_current_data</span><span class="p">[</span><span class="s">&#39;was_called&#39;</span><span class="p">]:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;NoCall&quot;</span>
                <span class="k">elif</span> <span class="ow">not</span> <span class="n">vcf_current_data</span><span class="p">[</span><span class="s">&#39;passed_coverage&#39;</span><span class="p">]:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;CovFail&quot;</span>
                <span class="k">elif</span> <span class="ow">not</span> <span class="n">vcf_current_data</span><span class="p">[</span><span class="s">&#39;passed_proportion&#39;</span><span class="p">]:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;PropFail&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;PASS&quot;</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span>
        <span class="c"># Determine if a line should be present in a particular matrix</span>
        <span class="c"># bestsnp outputs</span>
        <span class="k">if</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;indelcall&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">call_data</span><span class="p">[</span>
            <span class="s">&#39;refcall&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">genome_count</span> <span class="ow">or</span> <span class="n">dups_call</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">consensus_check</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;bestsnp&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;matrix&#39;</span> <span class="ow">or</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;vcf&#39;</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;fasta&#39;</span> <span class="ow">and</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;bestsnp&#39;</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">genome_identifier</span> <span class="ow">in</span> <span class="n">fasta_pending_data</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;fastadata&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append_contig</span><span class="p">(</span><span class="n">fasta_pending_data</span><span class="p">[</span><span class="n">genome_identifier</span><span class="p">],</span>
                                                                 <span class="n">genome_identifier</span><span class="p">)</span>
        <span class="c"># missing data outputs</span>
        <span class="k">if</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;indelcall&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">dups_call</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&quot;missingdata&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;matrix&#39;</span> <span class="ow">or</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;vcf&#39;</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;fasta&#39;</span> <span class="ow">and</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;missingdata&#39;</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">genome_identifier</span> <span class="ow">in</span> <span class="n">fasta_pending_data</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;fastadata&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append_contig</span><span class="p">(</span><span class="n">fasta_pending_data</span><span class="p">[</span><span class="n">genome_identifier</span><span class="p">],</span>
                                                                 <span class="n">genome_identifier</span><span class="p">)</span>
        <span class="c"># including ref outputs</span>
        <span class="k">if</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;indelcall&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">call_data</span><span class="p">[</span><span class="s">&#39;snpcall&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">call_data</span><span class="p">[</span>
            <span class="s">&#39;refcall&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">genome_count</span> <span class="ow">or</span> <span class="n">dups_call</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">consensus_check</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;includeref&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;matrix&#39;</span> <span class="ow">or</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;vcf&#39;</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;fasta&#39;</span> <span class="ow">and</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;includeref&#39;</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">genome_identifier</span> <span class="ow">in</span> <span class="n">fasta_pending_data</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;fastadata&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">append_contig</span><span class="p">(</span><span class="n">fasta_pending_data</span><span class="p">[</span><span class="n">genome_identifier</span><span class="p">],</span>
                                                                 <span class="n">genome_identifier</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">flush_cumulative_stat_cache</span><span class="p">()</span>

    <span class="c"># NOTE(jtravis): Fix documentation _write_matrix_line() does not exist</span>
<div class="viewcode-block" id="GenomeCollection.send_to_matrix_handles"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.send_to_matrix_handles">[docs]</a>    <span class="k">def</span> <span class="nf">send_to_matrix_handles</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrix_formats</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes headers and handles per-matrix logic.  Calls _write_matrix_line</span>
<span class="sd">        to handle the per-line computation and analysis.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;matrix&#39;</span><span class="p">:</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;LocusID</span><span class="se">\t</span><span class="s">Reference</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">identifier</span><span class="p">()))</span>
                <span class="k">for</span> <span class="n">genome_path</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_failed_genomes</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">genome_path</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;bestsnp&#39;</span> <span class="ow">or</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filter&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;includeref&#39;</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="s">&quot;#SNPcall</span><span class="se">\t</span><span class="s">#Indelcall</span><span class="se">\t</span><span class="s">#Refcall</span><span class="se">\t</span><span class="s">#CallWasMade</span><span class="se">\t</span><span class="s">#PassedDepthFilter</span><span class="se">\t</span><span class="s">#PassedProportionFilter</span><span class="se">\t</span><span class="s">#A</span><span class="se">\t</span><span class="s">#C</span><span class="se">\t</span><span class="s">#G</span><span class="se">\t</span><span class="s">#T</span><span class="se">\t</span><span class="s">#Indel</span><span class="se">\t</span><span class="s">#NXdegen</span><span class="se">\t</span><span class="s">Contig</span><span class="se">\t</span><span class="s">Position</span><span class="se">\t</span><span class="s">InDupRegion</span><span class="se">\t</span><span class="s">SampleConsensus</span><span class="se">\t</span><span class="s">Pattern</span><span class="se">\t</span><span class="s">Pattern#</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c"># must be all callable or missing data</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="s">&quot;#SNPcall</span><span class="se">\t</span><span class="s">#Indelcall</span><span class="se">\t</span><span class="s">#Refcall</span><span class="se">\t</span><span class="s">#CallWasMade</span><span class="se">\t</span><span class="s">#PassedDepthFilter</span><span class="se">\t</span><span class="s">#PassedProportionFilter</span><span class="se">\t</span><span class="s">#A</span><span class="se">\t</span><span class="s">#C</span><span class="se">\t</span><span class="s">#G</span><span class="se">\t</span><span class="s">#T</span><span class="se">\t</span><span class="s">#Indel</span><span class="se">\t</span><span class="s">#NXdegen</span><span class="se">\t</span><span class="s">&quot;</span> <span class="o">+</span>
                            <span class="s">&quot;Contig</span><span class="se">\t</span><span class="s">Position</span><span class="se">\t</span><span class="s">InDupRegion</span><span class="se">\t</span><span class="s">SampleConsensus</span><span class="se">\t</span><span class="s">CallWasMade</span><span class="se">\t</span><span class="s">PassedDepthFilter</span><span class="se">\t</span><span class="s">PassedProportionFilter</span><span class="se">\t</span><span class="s">Pattern</span><span class="se">\t</span><span class="s">Pattern#</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
            <span class="k">elif</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;fasta&#39;</span><span class="p">:</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;fastadata&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">GenomeStatus</span><span class="p">()</span>
                <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;fastadata&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">add_contig</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">identifier</span><span class="p">())</span>
            <span class="k">elif</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;vcf&#39;</span><span class="p">:</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##fileFormat=VCFv4.2</span><span class="se">\n</span><span class="s">##source=NASPv{0}</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">__version__</span><span class="p">))</span>
                <span class="k">for</span> <span class="n">current_contig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">get_contigs</span><span class="p">():</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##contig=&lt;ID=</span><span class="se">\&quot;</span><span class="s">{0}</span><span class="se">\&quot;</span><span class="s">,length={1}&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_contig</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">get_contig_length</span><span class="p">(</span><span class="n">current_contig</span><span class="p">)))</span>
                <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##SAMPLE=&lt;ID=</span><span class="se">\&quot;</span><span class="s">{0}</span><span class="se">\&quot;</span><span class="s">,Genomes=</span><span class="se">\&quot;</span><span class="s">{0}</span><span class="se">\&quot;</span><span class="s">,Mixture=1.0&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">identifier</span><span class="p">()))</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##INFO=&lt;ID=NS,Number=1,Type=Integer,Description=</span><span class="se">\&quot;</span><span class="s">Number of Samples With Data</span><span class="se">\&quot;</span><span class="s">&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##FILTER=&lt;ID=NoCall,Description=</span><span class="se">\&quot;</span><span class="s">No call for this sample at this position</span><span class="se">\&quot;</span><span class="s">&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##FILTER=&lt;ID=CovFail,Description=</span><span class="se">\&quot;</span><span class="s">Insufficient depth of coverage for this sample at this position</span><span class="se">\&quot;</span><span class="s">&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##FILTER=&lt;ID=PropFail,Description=</span><span class="se">\&quot;</span><span class="s">Insufficient proportion of reads were variant for this sample at this position</span><span class="se">\&quot;</span><span class="s">&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=</span><span class="se">\&quot;</span><span class="s">Genotype</span><span class="se">\&quot;</span><span class="s">&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;##FORMAT=&lt;ID=FT,Number=1,Type=String,Description=</span><span class="se">\&quot;</span><span class="s">Filters that failed for this sample at this position</span><span class="se">\&quot;</span><span class="s">&gt;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;#CHROM</span><span class="se">\t</span><span class="s">POS</span><span class="se">\t</span><span class="s">ID</span><span class="se">\t</span><span class="s">REF</span><span class="se">\t</span><span class="s">ALT</span><span class="se">\t</span><span class="s">QUAL</span><span class="se">\t</span><span class="s">FILTER</span><span class="se">\t</span><span class="s">INFO</span><span class="se">\t</span><span class="s">FORMAT&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">genome</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_genomes</span><span class="p">:</span>
                    <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span> <span class="o">+</span> <span class="n">genome</span><span class="o">.</span><span class="n">identifier</span><span class="p">())</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
        <span class="c"># Key None stores next unused</span>
        <span class="n">pattern_data</span> <span class="o">=</span> <span class="p">{</span><span class="bp">None</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>
        <span class="k">for</span> <span class="n">current_contig</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_contigs</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">current_pos</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_reference</span><span class="o">.</span><span class="n">get_contig_length</span><span class="p">(</span><span class="n">current_contig</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_format_matrix_line</span><span class="p">(</span><span class="n">current_contig</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">matrix_formats</span><span class="p">,</span> <span class="n">pattern_data</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;matrix&#39;</span> <span class="ow">and</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">])</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
                    <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;vcf&#39;</span> <span class="ow">and</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">])</span>
                        <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;linetowrite&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;dataformat&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;fasta&#39;</span><span class="p">:</span>
                <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;fastadata&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">send_to_fasta_handle</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">])</span>
</div>
<div class="viewcode-block" id="GenomeCollection.write_to_matrices"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.write_to_matrices">[docs]</a>    <span class="k">def</span> <span class="nf">write_to_matrices</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrix_formats</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Opens files for writing; abstracted for flexibility/testing. &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;filename&#39;</span><span class="p">],</span> <span class="s">&#39;w&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">send_to_matrix_handles</span><span class="p">(</span><span class="n">matrix_formats</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">matrix_format</span> <span class="ow">in</span> <span class="n">matrix_formats</span><span class="p">:</span>
            <span class="n">matrix_format</span><span class="p">[</span><span class="s">&#39;handle&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
</div>
    <span class="k">def</span> <span class="nf">_write_general_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">general_handle</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Writes finalized general stats to file. &quot;&quot;&quot;</span>
        <span class="n">denominator_stat</span> <span class="o">=</span> <span class="s">&#39;reference_length&#39;</span>
        <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Contig</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">current_stat</span> <span class="ow">in</span> <span class="n">general_stat_array</span><span class="p">:</span>
            <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_stat</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">current_stat</span> <span class="o">!=</span> <span class="n">denominator_stat</span><span class="p">:</span>
                <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0} (%)</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_stat</span><span class="p">))</span>
        <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
        <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">stat descriptions go here</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span> <span class="c"># FIXME</span>
        <span class="k">for</span> <span class="n">current_contig</span> <span class="ow">in</span> <span class="p">(</span> <span class="p">[</span><span class="bp">None</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_contigs</span><span class="p">()</span> <span class="p">):</span>
            <span class="n">denominator_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_contig_stat</span><span class="p">(</span><span class="n">denominator_stat</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">current_contig</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Whole Genome</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_contig</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">current_stat</span> <span class="ow">in</span> <span class="n">general_stat_array</span><span class="p">:</span>
                <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_contig_stat</span><span class="p">(</span><span class="n">current_stat</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">))))</span>
                <span class="k">if</span> <span class="n">current_stat</span> <span class="o">!=</span> <span class="n">denominator_stat</span><span class="p">:</span>
                    <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                        <span class="s">&quot;</span><span class="si">%.2f%%</span><span class="se">\t</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_contig_stat</span><span class="p">(</span><span class="n">current_stat</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span> <span class="o">/</span> <span class="n">denominator_value</span> <span class="o">*</span> <span class="mi">100</span> <span class="p">))</span>
            <span class="n">general_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_write_sample_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sample_handle</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Writes finalized sample stats to file.</span>
<span class="sd">        Includes per-sample counts, any/all counts across each set of</span>
<span class="sd">        sample-analyses, and any/all counts overall.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sample_stat_array</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;was_called&#39;</span><span class="p">,</span> <span class="s">&#39;passed_coverage_filter&#39;</span><span class="p">,</span> <span class="s">&#39;passed_proportion_filter&#39;</span><span class="p">,</span> <span class="s">&#39;quality_breadth&#39;</span><span class="p">,</span> 
                             <span class="s">&#39;called_reference&#39;</span><span class="p">,</span> <span class="s">&#39;called_snp&#39;</span><span class="p">,</span> <span class="s">&#39;called_degen&#39;</span><span class="p">]</span>
        <span class="n">denominator_stat</span> <span class="o">=</span> <span class="s">&#39;reference_length&#39;</span>
        <span class="n">denominator_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_contig_stat</span><span class="p">(</span><span class="n">denominator_stat</span><span class="p">)</span>
        <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;Sample</span><span class="se">\t</span><span class="s">Sample::Analysis</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">current_stat</span> <span class="ow">in</span> <span class="n">sample_stat_array</span><span class="p">:</span>
            <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_stat</span><span class="p">))</span>
            <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0} (%)</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_stat</span><span class="p">))</span>
        <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
        <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">stat descriptions go here</span><span class="se">\n\n</span><span class="s">&quot;</span><span class="p">)</span> <span class="c"># FIXME</span>
        <span class="k">for</span> <span class="n">current_sample</span> <span class="ow">in</span> <span class="p">(</span> <span class="p">[</span><span class="bp">None</span><span class="p">]</span> <span class="o">+</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_identifiers</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="p">):</span>
            <span class="k">for</span> <span class="n">current_analysis</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&#39;any&#39;</span><span class="p">,</span> <span class="s">&#39;all&#39;</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">current_sample</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                    <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_sample</span><span class="p">))</span>
                <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;[{0}]</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_analysis</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">current_sample</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                    <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">current_stat</span> <span class="ow">in</span> <span class="n">sample_stat_array</span><span class="p">:</span>
                    <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_cumulative_stat</span><span class="p">(</span><span class="n">current_stat</span><span class="p">,</span> <span class="n">current_analysis</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">))))</span>
                    <span class="k">if</span> <span class="n">current_stat</span> <span class="o">!=</span> <span class="n">denominator_stat</span><span class="p">:</span>
                        <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%.2f%%</span><span class="se">\t</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_cumulative_stat</span><span class="p">(</span><span class="n">current_stat</span><span class="p">,</span> <span class="n">current_analysis</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">)</span> <span class="o">/</span> <span class="n">denominator_value</span> <span class="o">*</span> <span class="mi">100</span><span class="p">))</span>
                <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">current_sample</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">current_analysis</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_genome_identifiers</span><span class="p">[</span><span class="n">current_sample</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
                    <span class="p">(</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span> <span class="p">)</span> <span class="o">=</span> <span class="n">current_analysis</span>
                    <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">current_sample</span><span class="p">,</span> <span class="n">sample_identifier</span><span class="p">))</span>
                    <span class="k">for</span> <span class="n">current_stat</span> <span class="ow">in</span> <span class="n">sample_stat_array</span><span class="p">:</span>
                        <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;{0}</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_sample_stat</span><span class="p">(</span><span class="n">current_stat</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">,</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span><span class="p">))))</span>
                        <span class="k">if</span> <span class="n">current_stat</span> <span class="o">!=</span> <span class="n">denominator_stat</span><span class="p">:</span>
                            <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%.2f%%</span><span class="se">\t</span><span class="s">&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get_sample_stat</span><span class="p">(</span><span class="n">current_stat</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">,</span> <span class="n">sample_identifier</span><span class="p">,</span> <span class="n">sample_path</span><span class="p">)</span> <span class="o">/</span> <span class="n">denominator_value</span> <span class="o">*</span> <span class="mi">100</span><span class="p">))</span>
                    <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="n">sample_handle</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>

<div class="viewcode-block" id="GenomeCollection.write_to_stats_files"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.GenomeCollection.write_to_stats_files">[docs]</a>    <span class="k">def</span> <span class="nf">write_to_stats_files</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">general_filename</span><span class="p">,</span> <span class="n">sample_filename</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Opens files for writing; abstracted for flexibility/testing. &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">general_filename</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">general_handle</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">sample_filename</span><span class="p">,</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">sample_handle</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_general_stats</span><span class="p">(</span><span class="n">general_handle</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_write_sample_stats</span><span class="p">(</span><span class="n">sample_handle</span><span class="p">)</span>
        <span class="c"># print( self._stats._contig_stats )</span>
        <span class="c"># print( self._stats._sample_stats )</span>

</div></div>
<div class="viewcode-block" id="VCFRecord"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord">[docs]</a><span class="k">class</span> <span class="nc">VCFRecord</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; VCF parser, object representing an input VCF being read. &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">file_path</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_file_path</span> <span class="o">=</span> <span class="n">file_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_file_handle</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_file_path</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_header_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_get_header_map</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">_get_header_map</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Pulls the headers from the file, so that we know which columns mean</span>
<span class="sd">        what as we read the file in line-by-line.</span>
<span class="sd">        The header map is expected to start with &quot;#CHROM&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">current_line</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
        <span class="k">while</span> <span class="n">current_line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">6</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&quot;#CHROM&quot;</span><span class="p">:</span>
            <span class="n">current_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_file_handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">current_line</span> <span class="o">==</span> <span class="s">&#39;&#39;</span><span class="p">:</span>
                <span class="k">raise</span> <span class="n">MalformedInputFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_file_path</span><span class="p">,</span> <span class="s">&quot;mandatory VCF header not found or not recognized&quot;</span><span class="p">)</span>
        <span class="n">header_list</span> <span class="o">=</span> <span class="n">current_line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
        <span class="n">sample_headers_started</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">for</span> <span class="n">current_header</span> <span class="ow">in</span> <span class="n">header_list</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_header_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_header</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">sample_headers_started</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_header</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">current_header</span> <span class="o">==</span> <span class="s">&quot;FORMAT&quot;</span><span class="p">:</span>
                <span class="n">sample_headers_started</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">sample_headers_started</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s">&#39;vcf_sample&#39;</span><span class="p">)</span>
            <span class="c"># print( self._header_list )</span>
            <span class="c"># print( self._sample_list )</span>

    <span class="k">def</span> <span class="nf">_get_record_alts</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get and parse the list of non-reference calls that might appear in</span>
<span class="sd">        one or more samples.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;alts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;REF&#39;</span><span class="p">]]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;ALT&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&#39;.&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;alts&#39;</span><span class="p">]</span> <span class="o">+=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;ALT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_get_record_info</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">info_strings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;INFO&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;;&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">info_string</span> <span class="ow">in</span> <span class="n">info_strings</span><span class="p">:</span>
            <span class="n">info_keyval</span> <span class="o">=</span> <span class="n">info_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">info_keyval</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="n">info_keyval</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">info_keyval</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="n">info_keyval</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="bp">None</span>

    <span class="k">def</span> <span class="nf">_get_record_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="s">&#39;FORMAT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">sample_header</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;FORMAT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">current_sample</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
                    <span class="nb">zip</span><span class="p">(</span><span class="n">sample_header</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)))</span>

<div class="viewcode-block" id="VCFRecord.fetch_next_record"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.fetch_next_record">[docs]</a>    <span class="k">def</span> <span class="nf">fetch_next_record</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        We&#39;re done with the current line, let&#39;s move to the next.</span>
<span class="sd">        Populates all the information at the position we&#39;ll need for parsing</span>
<span class="sd">        parts of the next line.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">current_line</span> <span class="o">=</span> <span class="s">&quot;#&quot;</span>
        <span class="k">while</span> <span class="n">current_line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s">&quot;#&quot;</span><span class="p">:</span>
            <span class="n">current_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_file_handle</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">if</span> <span class="n">current_line</span> <span class="o">!=</span> <span class="s">&#39;&#39;</span><span class="p">:</span>
            <span class="n">record_list</span> <span class="o">=</span> <span class="n">current_line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_header_list</span><span class="p">,</span> <span class="n">record_list</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_get_record_alts</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_get_record_info</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_get_record_samples</span><span class="p">()</span>
            <span class="n">return_value</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="c"># print( self._current_record )</span>
        <span class="k">return</span> <span class="n">return_value</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_samples"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_samples">[docs]</a>    <span class="k">def</span> <span class="nf">get_samples</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_contig"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_contig">[docs]</a>    <span class="k">def</span> <span class="nf">get_contig</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;CHROM&#39;</span><span class="p">]</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_position"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_position">[docs]</a>    <span class="k">def</span> <span class="nf">get_position</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;POS&#39;</span><span class="p">])</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_reference_call"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_reference_call">[docs]</a>    <span class="k">def</span> <span class="nf">get_reference_call</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;REF&#39;</span><span class="p">]</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_sample_call"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_sample_call">[docs]</a>    <span class="k">def</span> <span class="nf">get_sample_call</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">):</span>
        <span class="c"># FIXME indels</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;alts&#39;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;alts&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">elif</span> <span class="s">&#39;FORMAT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="s">&#39;GT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]:</span>
                <span class="n">alt_number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;GT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">alt_number</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
                    <span class="n">return_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;alts&#39;</span><span class="p">][</span><span class="nb">int</span><span class="p">(</span><span class="n">alt_number</span><span class="p">)]</span>
                    <span class="c"># OMG varscan</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;REF&#39;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;REF&#39;</span><span class="p">])</span> <span class="o">-</span> <span class="mi">1</span> <span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">return_value</span><span class="p">)</span> <span class="ow">and</span> \
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;REF&#39;</span><span class="p">][:</span><span class="nb">len</span><span class="p">(</span><span class="n">return_value</span><span class="p">)]</span> <span class="o">!=</span> <span class="n">return_value</span> <span class="ow">and</span> \
                                    <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;REF&#39;</span><span class="p">][</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">return_value</span><span class="p">):]</span> <span class="o">==</span> <span class="n">return_value</span><span class="p">:</span>
                        <span class="n">return_value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;alts&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="c">#elif len(self._current_record[&#39;alts&#39;]) &gt; 1:</span>
        <span class="c">#    return_value = self._current_record[&#39;alts&#39;][1]</span>
        <span class="k">return</span> <span class="n">return_value</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_coverage"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_coverage">[docs]</a>    <span class="k">def</span> <span class="nf">get_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">):</span>
        <span class="n">sample_coverage</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="s">&#39;FORMAT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="s">&#39;DP&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]</span> <span class="ow">and</span> \
                        <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;DP&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> \
                <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;DP&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
            <span class="n">sample_coverage</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;DP&#39;</span><span class="p">])</span>
        <span class="k">elif</span> <span class="s">&#39;DP&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;DP&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> \
                <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;DP&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
            <span class="n">sample_coverage</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;DP&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span><span class="p">)</span>
        <span class="k">elif</span> <span class="s">&#39;ADP&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;ADP&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> \
                <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;ADP&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
            <span class="n">sample_coverage</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;ADP&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span><span class="p">)</span>
        <span class="c"># NASP output</span>
        <span class="k">elif</span> <span class="s">&#39;FORMAT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="s">&#39;FT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]:</span>
            <span class="n">failed_filters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;FT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s">&#39;CovFail&#39;</span> <span class="ow">in</span> <span class="n">failed_filters</span><span class="p">:</span>
                <span class="n">sample_coverage</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">elif</span> <span class="s">&#39;PASS&#39;</span> <span class="ow">in</span> <span class="n">failed_filters</span> <span class="ow">or</span> <span class="s">&#39;PropFail&#39;</span> <span class="ow">in</span> <span class="n">failed_filters</span><span class="p">:</span>
                <span class="n">sample_coverage</span> <span class="o">=</span> <span class="s">&#39;PASS&#39;</span>
        <span class="k">return</span> <span class="n">sample_coverage</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_proportion"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_proportion">[docs]</a>    <span class="k">def</span> <span class="nf">get_proportion</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">,</span> <span class="n">sample_coverage</span><span class="p">,</span> <span class="n">is_a_snp</span><span class="p">):</span>
        <span class="n">sample_proportion</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="k">if</span> <span class="s">&#39;FORMAT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="s">&#39;AD&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]:</span>
            <span class="n">call_depths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;AD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
            <span class="c"># gatk, reliable and documented</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">call_depths</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">alt_number</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;GT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;|&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">alt_number</span><span class="o">.</span><span class="n">isdigit</span><span class="p">():</span>
                    <span class="n">sample_proportion</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">call_depths</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">alt_number</span><span class="p">)])</span> <span class="o">/</span> <span class="n">sample_coverage</span>
                    <span class="c"># varscan, reliable and documented</span>
            <span class="k">elif</span> <span class="n">is_a_snp</span><span class="p">:</span>
                <span class="n">sample_proportion</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">call_depths</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">/</span> <span class="n">sample_coverage</span>
            <span class="k">elif</span> <span class="ow">not</span> <span class="n">is_a_snp</span> <span class="ow">and</span> <span class="s">&#39;RD&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]:</span>
                <span class="n">sample_proportion</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;RD&#39;</span><span class="p">])</span> <span class="o">/</span> <span class="n">sample_coverage</span>
        <span class="c"># solsnp, undocumented, no multi-sample support</span>
        <span class="k">elif</span> <span class="s">&#39;AR&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">]:</span>
            <span class="n">sample_proportion</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;AR&#39;</span><span class="p">])</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">is_a_snp</span><span class="p">:</span>
                <span class="n">sample_proportion</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">sample_proportion</span>
        <span class="c"># samtools, estimate, dubious accuracy</span>
        <span class="k">elif</span> <span class="s">&#39;DP4&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">]:</span>
            <span class="n">call_depths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;info&#39;</span><span class="p">][</span><span class="s">&#39;DP4&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">is_a_snp</span><span class="p">:</span>
                <span class="n">sample_proportion</span> <span class="o">=</span> <span class="p">(</span> <span class="nb">int</span><span class="p">(</span><span class="n">call_depths</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">call_depths</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span> <span class="p">)</span> <span class="o">/</span> <span class="p">(</span>
                    <span class="n">sample_coverage</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span><span class="p">)</span> <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sample_proportion</span> <span class="o">=</span> <span class="p">(</span> <span class="nb">int</span><span class="p">(</span><span class="n">call_depths</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">call_depths</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="p">)</span> <span class="o">/</span> <span class="p">(</span>
                    <span class="n">sample_coverage</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_sample_list</span><span class="p">)</span> <span class="p">)</span>
        <span class="c"># NASP output</span>
        <span class="k">elif</span> <span class="s">&#39;FORMAT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="s">&#39;FT&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">]:</span>
            <span class="n">failed_filters</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;samples&#39;</span><span class="p">][</span><span class="n">current_sample</span><span class="p">][</span><span class="s">&#39;FT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s">&#39;PropFail&#39;</span> <span class="ow">in</span> <span class="n">failed_filters</span><span class="p">:</span>
                <span class="n">sample_proportion</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
            <span class="k">elif</span> <span class="s">&#39;PASS&#39;</span> <span class="ow">in</span> <span class="n">failed_filters</span><span class="p">:</span>
                <span class="n">sample_proportion</span> <span class="o">=</span> <span class="s">&#39;PASS&#39;</span>
        <span class="k">return</span> <span class="n">sample_proportion</span>
</div>
<div class="viewcode-block" id="VCFRecord.get_sample_info"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.VCFRecord.get_sample_info">[docs]</a>    <span class="k">def</span> <span class="nf">get_sample_info</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">current_sample</span><span class="p">):</span>
        <span class="n">sample_info</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;call&#39;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_sample_call</span><span class="p">(</span><span class="n">current_sample</span><span class="p">)}</span>
        <span class="c"># FIXME indels</span>
        <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;was_called&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;is_a_snp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">False</span>
        <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span><span class="p">:</span>
            <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;was_called&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>
            <span class="k">if</span> <span class="n">Genome</span><span class="o">.</span><span class="n">simple_call</span><span class="p">(</span><span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">])</span> <span class="o">!=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">simple_call</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_current_record</span><span class="p">[</span><span class="s">&#39;global&#39;</span><span class="p">][</span><span class="s">&#39;REF&#39;</span><span class="p">]):</span>
                <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;is_a_snp&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="c"># FIXME indels</span>
        <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;is_an_insert&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;is_a_delete&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">None</span>
        <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_coverage</span><span class="p">(</span><span class="n">current_sample</span><span class="p">)</span>
        <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;proportion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_proportion</span><span class="p">(</span><span class="n">current_sample</span><span class="p">,</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;coverage&#39;</span><span class="p">],</span>
                                                        <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;is_a_snp&#39;</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">sample_info</span>

</div></div>
<div class="viewcode-block" id="InvalidContigName"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.InvalidContigName">[docs]</a><span class="k">class</span> <span class="nc">InvalidContigName</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A contig was referenced that is not known to exist in this run.</span>
<span class="sd">    For the most part, this exception isn&#39;t used because encountering</span>
<span class="sd">    an unknown contig either means we should create it or we should</span>
<span class="sd">    discard the data that follows.  However, there are potential times</span>
<span class="sd">    where we need to take more extreme action.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">invalid_contig</span><span class="p">,</span> <span class="n">contig_list</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_invalid_contig</span> <span class="o">=</span> <span class="n">invalid_contig</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig_list</span> <span class="o">=</span> <span class="n">contig_list</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s">&quot;Contig &#39;{0}&#39; not in {1}!&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_invalid_contig</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_contig_list</span><span class="p">))</span>

</div>
<div class="viewcode-block" id="ReferenceCallMismatch"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.ReferenceCallMismatch">[docs]</a><span class="k">class</span> <span class="nc">ReferenceCallMismatch</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    At two different points in the run, the reference call appeared, and the</span>
<span class="sd">    two calls disagreed.  This probably means the user is analyzing two</span>
<span class="sd">    different sets of data as if they belong together, in error.  This is</span>
<span class="sd">    a very bad thing.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">old_call</span><span class="p">,</span> <span class="n">new_call</span><span class="p">,</span> <span class="n">data_file</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">contig_name</span><span class="o">=</span><span class="bp">None</span><span class="p">,</span> <span class="n">position_number</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_old_call</span> <span class="o">=</span> <span class="n">old_call</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_new_call</span> <span class="o">=</span> <span class="n">new_call</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_data_file</span> <span class="o">=</span> <span class="n">data_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_contig_name</span> <span class="o">=</span> <span class="n">contig_name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_position_number</span> <span class="o">=</span> <span class="n">position_number</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        The file, contig, and position are optional.  When given in any</span>
<span class="sd">        combination, what&#39;s given will be in the error message.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="s">&quot;Expected reference call of &#39;{0}&#39; but got &#39;{1}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_old_call</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_new_call</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">+=</span> <span class="s">&quot; while reading &#39;{0}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_position_number</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">+=</span> <span class="s">&quot; at position {0}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_position_number</span><span class="p">))</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_contig_name</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">+=</span> <span class="s">&quot; on contig &#39;{0}&#39;&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_contig_name</span><span class="p">)</span>
        <span class="n">return_value</span> <span class="o">+=</span> <span class="s">&quot;!&quot;</span>
        <span class="k">return</span> <span class="n">return_value</span>

</div>
<div class="viewcode-block" id="MalformedInputFile"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.MalformedInputFile">[docs]</a><span class="k">class</span> <span class="nc">MalformedInputFile</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    There was an error parsing the input file.</span>
<span class="sd">    Give the information we can to the user in the stderr stream.</span>
<span class="sd">    Let&#39;s hope they read it.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">data_file</span><span class="p">,</span> <span class="n">error_message</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_data_file</span> <span class="o">=</span> <span class="n">data_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_error_message</span> <span class="o">=</span> <span class="n">error_message</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">return_value</span> <span class="o">=</span> <span class="s">&quot;Input file &#39;{0}&#39; seems to be malformed&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_error_message</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">return_value</span> <span class="o">+=</span> <span class="s">&quot;: {0}&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_error_message</span><span class="p">)</span>
        <span class="n">return_value</span> <span class="o">+=</span> <span class="s">&quot;!&quot;</span>
        <span class="k">return</span> <span class="n">return_value</span>

</div>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../nasp.html#nasp.nasp_objects.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s">&quot;This is just a class definition, to be included from other python programs.&quot;</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s">&quot;  ...You sillyface.&quot;</span><span class="p">)</span>

</div>
<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, Darrin Lemmer.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.9.8',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>