

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>nasp.vcf_to_matrix &mdash; Northern Arizona SNP Pipeline 0.9.8 documentation</title>
  

  
  
    <link rel="shortcut icon" href="../../_static/TGenLogo.ico"/>
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  
    <link rel="top" title="Northern Arizona SNP Pipeline 0.9.8 documentation" href="../../index.html"/>
        <link rel="up" title="Module code" href="../index.html"/> 

  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="../../index.html" class="fa fa-home"> Northern Arizona SNP Pipeline</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">The Northern Arizona SNP Pipeline (NASP)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#getting-started">Getting Started</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#repeating-analysis">Repeating Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#license">License</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../readme.html#contact">Contact</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#install">Install</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#update">Update</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#uninstall">Uninstall</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#required-dependencies">Required Dependencies</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#optional-dependencies">Optional Dependencies</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../usage.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#nasp">nasp</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#format-fasta">format_fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#find-duplicates">find_duplicates</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../usage.html#vcf-to-matrix">vcf_to_matrix</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../scripts.html">Post-Analysis Scripts</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../scripts.html#filter-matrix-by-coord-py">filter_matrix_by_coord.py</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../scripts.html#filter-matrix-by-genome-py">filter_matrix_by_genome.py</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../scripts.html#matrix-to-fasta-py">matrix_to_fasta.py</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../scripts.html#report-single-snps-single-isolate-py">report_single_snps_single_isolate.py</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../contributing.html">Contributing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#types-of-contributions">Types of Contributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#get-started">Get Started!</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#pull-request-guidelines">Pull Request Guidelines</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../contributing.html#tips">Tips</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../authors.html">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../authors.html#development-lead">Development Lead</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../authors.html#contributors">Contributors</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../history.html#id1">0.9.6 (2014-06-05)</a></li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../index.html">Northern Arizona SNP Pipeline</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../index.html">Module code</a> &raquo;</li>
      
    <li>nasp.vcf_to_matrix</li>
      <li class="wy-breadcrumbs-aside">
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  <h1>Source code for nasp.vcf_to_matrix</h1><div class="highlight"><pre>
<span class="c">#!/usr/bin/env python3</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s">&quot;David Smith&quot;</span>
<span class="n">__version__</span> <span class="o">=</span> <span class="s">&quot;0.9.8&quot;</span>
<span class="n">__email__</span> <span class="o">=</span> <span class="s">&quot;dsmith@tgen.org&quot;</span>

<span class="kn">import</span> <span class="nn">logging</span>


<span class="k">def</span> <span class="nf">_parse_args</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Parse command line arguments and return an object whose attributes are the</span>
<span class="sd">    settings provided, as per standard argparse behavior.</span>
<span class="sd">    For backward compatibility when we were changing dispatchers, the options</span>
<span class="sd">    could be directly provided on the command line, or a single argument could</span>
<span class="sd">    specify the path to the XML file with the remaining arguments in it.  As</span>
<span class="sd">    the old dispatcher is deprecated, so perhaps is every option below except</span>
<span class="sd">    &#39;--dto-file&#39;, which would then be made required.</span>

<span class="sd">    Returns:</span>
<span class="sd">        argparse.Namespace: The properties of this object are the parsed command line arguments</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">argparse</span>

    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s">&quot;Meant to be called from the pipeline automatically.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--mode&quot;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;commandline&#39;</span><span class="p">,</span> <span class="s">&#39;xml&#39;</span><span class="p">],</span>
                        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Data passing mode, must be set to &#39;commandline&#39; or &#39;xml&#39;.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--reference-fasta&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Path to input reference fasta file.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--reference-dups&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Path to input reference dups file.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--input-files&quot;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s">&quot;+&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Path to input VCF/fasta files for matrix conversion.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--matrix-folder&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;matrices&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Name of folder to write output matrices to.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--stats-folder&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s">&quot;statistics&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Name of folder to write statistics files to.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--minimum-coverage&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Minimum coverage depth at a position.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--minimum-proportion&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s">&quot;Minimum proportion of reads that must match the call at a position.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--num-threads&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Number of threads to use when processing input.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s">&quot;--dto-file&quot;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s">&quot;Path to a matrix_dto XML file that defines all the parameters.&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>


<span class="k">def</span> <span class="nf">_parse_input_config</span><span class="p">(</span><span class="n">commandline_args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Set the configuration options from the XML file and add them as attributes</span>
<span class="sd">    to the object returned by _parse_args().</span>
<span class="sd">    Since getting these options directly from the command line is perhaps</span>
<span class="sd">    deprecated, we should probably store the configuration options from the</span>
<span class="sd">    XML file in a separate dictionary, instead of mixing the two data sources</span>
<span class="sd">    together as now.</span>

<span class="sd">    Args:</span>
<span class="sd">        commandline_args (argparse.Namespace):</span>

<span class="sd">    Returns:</span>
<span class="sd">        argparse.Namespace:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">nasp.matrix_DTO</span> <span class="kn">as</span> <span class="nn">matrix_DTO</span>

    <span class="p">(</span><span class="n">matrix_parms</span><span class="p">,</span> <span class="n">input_files</span><span class="p">)</span> <span class="o">=</span> <span class="n">matrix_DTO</span><span class="o">.</span><span class="n">parse_dto</span><span class="p">(</span><span class="n">commandline_args</span><span class="o">.</span><span class="n">dto_file</span><span class="p">)</span>
    <span class="n">commandline_args</span><span class="o">.</span><span class="n">reference_fasta</span> <span class="o">=</span> <span class="n">matrix_parms</span><span class="p">[</span><span class="s">&#39;reference-fasta&#39;</span><span class="p">]</span>
    <span class="n">commandline_args</span><span class="o">.</span><span class="n">reference_dups</span> <span class="o">=</span> <span class="n">matrix_parms</span><span class="p">[</span><span class="s">&#39;reference-dups&#39;</span><span class="p">]</span>
    <span class="n">commandline_args</span><span class="o">.</span><span class="n">matrix_folder</span> <span class="o">=</span> <span class="n">matrix_parms</span><span class="p">[</span><span class="s">&#39;matrix-folder&#39;</span><span class="p">]</span>
    <span class="n">commandline_args</span><span class="o">.</span><span class="n">stats_folder</span> <span class="o">=</span> <span class="n">matrix_parms</span><span class="p">[</span><span class="s">&#39;stats-folder&#39;</span><span class="p">]</span>
    <span class="n">commandline_args</span><span class="o">.</span><span class="n">minimum_coverage</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span>
        <span class="n">matrix_parms</span><span class="p">[</span><span class="s">&#39;minimum-coverage&#39;</span><span class="p">])</span> <span class="k">if</span> <span class="s">&quot;minimum-coverage&quot;</span> <span class="ow">in</span> <span class="n">matrix_parms</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="n">commandline_args</span><span class="o">.</span><span class="n">minimum_proportion</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span>
        <span class="n">matrix_parms</span><span class="p">[</span><span class="s">&#39;minimum-proportion&#39;</span><span class="p">])</span> <span class="k">if</span> <span class="s">&quot;minimum-proportion&quot;</span> <span class="ow">in</span> <span class="n">matrix_parms</span> <span class="k">else</span> <span class="mi">0</span>
    <span class="k">if</span> <span class="s">&quot;filter-matrix-format&quot;</span> <span class="ow">in</span> <span class="n">matrix_parms</span><span class="p">:</span>
        <span class="n">commandline_args</span><span class="o">.</span><span class="n">filter_matrix_format</span> <span class="o">=</span> <span class="n">matrix_parms</span><span class="p">[</span><span class="s">&#39;filter-matrix-format&#39;</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">commandline_args</span><span class="o">.</span><span class="n">filter_matrix_format</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">commandline_args</span><span class="o">.</span><span class="n">input_files</span> <span class="o">=</span> <span class="n">input_files</span>
    <span class="k">return</span> <span class="n">commandline_args</span>


<div class="viewcode-block" id="import_reference"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.import_reference">[docs]</a><span class="k">def</span> <span class="nf">import_reference</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">reference_path</span><span class="p">,</span> <span class="n">dups_path</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Take an empty reference object and populate it with the data from a</span>
<span class="sd">    reference file, and a dups file if any.</span>
<span class="sd">    Does not return anything, as the passed-in object is modified.</span>

<span class="sd">    Args:</span>
<span class="sd">        reference:</span>
<span class="sd">        reference_path:</span>
<span class="sd">        dups_path:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">reference</span><span class="o">.</span><span class="n">import_fasta_file</span><span class="p">(</span><span class="n">reference_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">dups_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="n">reference</span><span class="o">.</span><span class="n">import_dups_file</span><span class="p">(</span><span class="n">dups_path</span><span class="p">)</span>
        <span class="c"># from sys import stdout</span>
        <span class="c">#reference._genome._send_to_fasta_handle( stdout )</span>
        <span class="c">#reference._dups._send_to_fasta_handle( stdout )</span>

</div>
<div class="viewcode-block" id="import_external_fasta"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.import_external_fasta">[docs]</a><span class="k">def</span> <span class="nf">import_external_fasta</span><span class="p">(</span><span class="n">input_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a FastaGenome object, set its metadata, and populate it with the</span>
<span class="sd">    data from a fasta file.</span>
<span class="sd">    Must return the data as the single item in an array, because other file</span>
<span class="sd">    formats potentially contain several genomes.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_file:</span>

<span class="sd">    Returns:</span>
<span class="sd">        list of FastaGenome:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">nasp.nasp_objects</span> <span class="kn">import</span> <span class="n">FastaGenome</span>

    <span class="n">genome</span> <span class="o">=</span> <span class="n">FastaGenome</span><span class="p">()</span>
    <span class="n">set_genome_metadata</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">input_file</span><span class="p">)</span>
    <span class="n">genome</span><span class="o">.</span><span class="n">import_fasta_file</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">file_path</span><span class="p">(),</span> <span class="s">&quot;franken::&quot;</span><span class="p">)</span>
    <span class="c"># from sys import stdout</span>
    <span class="c">#genome._genome._send_to_fasta_handle( stdout )</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">genome</span><span class="p">]</span>


<span class="c"># FIXME split into a larger number of smaller more testable functions</span>
<span class="c"># FIXME This belongs in VCFGenome object perhaps?</span></div>
<div class="viewcode-block" id="read_vcf_file"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.read_vcf_file">[docs]</a><span class="k">def</span> <span class="nf">read_vcf_file</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">min_coverage</span><span class="p">,</span> <span class="n">min_proportion</span><span class="p">,</span> <span class="n">input_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Submit VCF to be read in to VCF parser, populate genome data and filter</span>
<span class="sd">    data from the parsed VCF data, return a list of the read-in genomes.</span>

<span class="sd">    Args:</span>
<span class="sd">        reference:</span>
<span class="sd">        min_converage (int):</span>
<span class="sd">        min_proportion (int):</span>
<span class="sd">        input_file (str):</span>

<span class="sd">    Returns:</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genomes</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">file_path</span> <span class="o">=</span> <span class="n">get_file_path</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
    <span class="c"># NOTE(jtravis): unused variable vcf_filehandle</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">vcf_filehandle</span><span class="p">:</span>
        <span class="kn">from</span> <span class="nn">nasp.nasp_objects</span> <span class="kn">import</span> <span class="n">VCFGenome</span><span class="p">,</span> <span class="n">Genome</span><span class="p">,</span> <span class="n">ReferenceCallMismatch</span><span class="p">,</span> <span class="n">VCFRecord</span>

        <span class="n">vcf_record</span> <span class="o">=</span> <span class="n">VCFRecord</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
        <span class="n">vcf_samples</span> <span class="o">=</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">get_samples</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">vcf_sample</span> <span class="ow">in</span> <span class="n">vcf_samples</span><span class="p">:</span>
            <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span> <span class="o">=</span> <span class="n">VCFGenome</span><span class="p">()</span>
            <span class="n">set_genome_metadata</span><span class="p">(</span><span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">],</span> <span class="n">input_file</span><span class="p">)</span>
            <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_nickname</span><span class="p">(</span><span class="n">vcf_sample</span><span class="p">)</span>
        <span class="k">while</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">fetch_next_record</span><span class="p">():</span>
            <span class="n">current_contig</span> <span class="o">=</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">get_contig</span><span class="p">()</span>
            <span class="n">current_pos</span> <span class="o">=</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">get_position</span><span class="p">()</span>
            <span class="c"># Skip if position isn&#39;t in reference; maybe user truncated reference to exclude an uninteresting region.</span>
            <span class="k">if</span> <span class="n">current_pos</span> <span class="o">&lt;=</span> <span class="n">reference</span><span class="o">.</span><span class="n">get_contig_length</span><span class="p">(</span><span class="n">current_contig</span><span class="p">):</span>
                <span class="n">reference_call</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">get_call</span><span class="p">(</span><span class="n">current_pos</span><span class="p">,</span> <span class="bp">None</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
                <span class="n">simplified_refcall</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">simple_call</span><span class="p">(</span><span class="n">reference_call</span><span class="p">)</span>
                <span class="k">if</span> <span class="p">(</span> <span class="n">simplified_refcall</span> <span class="o">!=</span> <span class="s">&#39;N&#39;</span> <span class="p">)</span> <span class="ow">and</span> <span class="p">(</span>
                    <span class="n">simplified_refcall</span> <span class="o">!=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">simple_call</span><span class="p">(</span><span class="n">vcf_record</span><span class="o">.</span><span class="n">get_reference_call</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span> <span class="p">):</span>
                    <span class="c"># Reference call from reference fasta differs from reference call in VCF file at the same position.</span>
                    <span class="k">raise</span> <span class="n">ReferenceCallMismatch</span><span class="p">(</span><span class="n">reference_call</span><span class="p">,</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">get_reference_call</span><span class="p">(),</span> <span class="n">file_path</span><span class="p">,</span>
                                                <span class="n">current_contig</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">vcf_sample</span> <span class="ow">in</span> <span class="n">vcf_samples</span><span class="p">:</span>
                    <span class="n">sample_info</span> <span class="o">=</span> <span class="n">vcf_record</span><span class="o">.</span><span class="n">get_sample_info</span><span class="p">(</span><span class="n">vcf_sample</span><span class="p">)</span>
                    <span class="c"># FIXME indels</span>
                    <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                        <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_call</span><span class="p">(</span><span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;call&#39;</span><span class="p">],</span> <span class="n">current_pos</span><span class="p">,</span> <span class="s">&#39;X&#39;</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;was_called&#39;</span><span class="p">]:</span>
                        <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_was_called</span><span class="p">(</span><span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;coverage&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;PASS&#39;</span> <span class="ow">or</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;coverage&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_coverage</span><span class="p">:</span>
                            <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_coverage_pass</span><span class="p">(</span><span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_coverage_pass</span><span class="p">(</span><span class="s">&#39;N&#39;</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;proportion&#39;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;proportion&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s">&#39;PASS&#39;</span> <span class="ow">or</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;proportion&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_proportion</span><span class="p">:</span>
                            <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_proportion_pass</span><span class="p">(</span><span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_proportion_pass</span><span class="p">(</span><span class="s">&#39;N&#39;</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="ow">not</span> <span class="n">sample_info</span><span class="p">[</span><span class="s">&#39;is_a_snp&#39;</span><span class="p">]:</span>
                        <span class="c"># Some big SNP callers, like GATK, do not provide proportion information when</span>
                        <span class="c"># the position is called reference.  We cannot filter these positions.</span>
                        <span class="n">genomes</span><span class="p">[</span><span class="n">vcf_sample</span><span class="p">]</span><span class="o">.</span><span class="n">set_proportion_pass</span><span class="p">(</span><span class="s">&#39;-&#39;</span><span class="p">,</span> <span class="n">current_pos</span><span class="p">,</span> <span class="n">current_contig</span><span class="p">)</span>
    <span class="c"># from sys import stdout</span>
    <span class="c">#for genome in genomes:</span>
    <span class="c">#    genomes[genome]._genome._send_to_fasta_handle( stdout )</span>
    <span class="k">return</span> <span class="n">genomes</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>


<span class="c"># FIXME The information these three functions capture is in the XML file.</span>
<span class="c"># FIXME These should all be deprecated and the XML data used instead.</span></div>
<div class="viewcode-block" id="determine_file_type"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.determine_file_type">[docs]</a><span class="k">def</span> <span class="nf">determine_file_type</span><span class="p">(</span><span class="n">input_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the file type from the packed input filename string</span>

<span class="sd">    Args:</span>
<span class="sd">        input_file (str):</span>

<span class="sd">    Returns:</span>
<span class="sd">        str:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">re</span>

    <span class="n">file_type</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="c"># Format is &quot;type,aligner,caller,::path&quot;</span>
    <span class="n">filename_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^((?:[^,:]+,)+)::.*$&#39;</span><span class="p">,</span> <span class="n">input_file</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">filename_match</span><span class="p">:</span>
        <span class="n">generator_array</span> <span class="o">=</span> <span class="n">filename_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
        <span class="n">file_type</span> <span class="o">=</span> <span class="n">generator_array</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c"># print( file_type )</span>
    <span class="k">return</span> <span class="n">file_type</span>

</div>
<div class="viewcode-block" id="get_file_path"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.get_file_path">[docs]</a><span class="k">def</span> <span class="nf">get_file_path</span><span class="p">(</span><span class="n">input_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Get the file path from the packed input filename string</span>

<span class="sd">    Args:</span>
<span class="sd">        input_file (str):</span>

<span class="sd">    Returns:</span>
<span class="sd">        str:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">re</span>

    <span class="n">file_path</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="c"># Format is &quot;type,aligner,caller,::path&quot;</span>
    <span class="n">filename_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^(?:[^,:]+,)+::(.*)$&#39;</span><span class="p">,</span> <span class="n">input_file</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">filename_match</span><span class="p">:</span>
        <span class="n">file_path</span> <span class="o">=</span> <span class="n">filename_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">file_path</span> <span class="o">=</span> <span class="n">input_file</span>
    <span class="c"># print( file_path )</span>
    <span class="k">return</span> <span class="n">file_path</span>

</div>
<div class="viewcode-block" id="set_genome_metadata"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.set_genome_metadata">[docs]</a><span class="k">def</span> <span class="nf">set_genome_metadata</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">input_file</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Get all the metadata from the packed input filename string</span>

<span class="sd">    Args:</span>
<span class="sd">        genome:</span>
<span class="sd">        input_file:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">re</span>
    <span class="c"># Format is &quot;type,aligner,caller,::path&quot;</span>
    <span class="n">filename_match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s">r&#39;^((?:[^,:]+,)+)::(.*)$&#39;</span><span class="p">,</span> <span class="n">input_file</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">filename_match</span><span class="p">:</span>
        <span class="n">generator_array</span> <span class="o">=</span> <span class="n">filename_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;,&#39;</span><span class="p">)</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">set_file_type</span><span class="p">(</span><span class="n">generator_array</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">add_generators</span><span class="p">(</span><span class="n">generator_array</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">set_file_path</span><span class="p">(</span><span class="n">filename_match</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">genome</span><span class="o">.</span><span class="n">set_file_path</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
        <span class="c"># print( genome.identifier() )</span>

</div>
<div class="viewcode-block" id="manage_input_thread"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.manage_input_thread">[docs]</a><span class="k">def</span> <span class="nf">manage_input_thread</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">min_coverage</span><span class="p">,</span> <span class="n">min_proportion</span><span class="p">,</span> <span class="n">input_q</span><span class="p">,</span> <span class="n">output_q</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Manage one input file worker thread, for reading the data from the file.</span>
<span class="sd">    Input filenames are pulled one at a time from the input queue, and the</span>
<span class="sd">    genome data from the read-in files is placed on the output queue.</span>
<span class="sd">    When an input filename of &quot;None&quot; appears, we know we&#39;re done and put</span>
<span class="sd">    &quot;None&quot; on the output queue so the controlling thread knows we won&#39;t be</span>
<span class="sd">    adding more data.</span>

<span class="sd">    Args:</span>
<span class="sd">        reference:</span>
<span class="sd">        min_coverage:</span>
<span class="sd">        min_proportion:</span>
<span class="sd">        input_q:</span>
<span class="sd">        output_q:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">input_file</span> <span class="o">=</span> <span class="n">input_q</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>
    <span class="k">while</span> <span class="n">input_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">new_genomes</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">file_type</span> <span class="o">=</span> <span class="n">determine_file_type</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s">&quot;frankenfasta&quot;</span><span class="p">:</span>
                <span class="n">new_genomes</span> <span class="o">=</span> <span class="n">import_external_fasta</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">file_type</span> <span class="o">==</span> <span class="s">&quot;vcf&quot;</span><span class="p">:</span>
                <span class="n">new_genomes</span> <span class="o">=</span> <span class="n">read_vcf_file</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">min_coverage</span><span class="p">,</span> <span class="n">min_proportion</span><span class="p">,</span> <span class="n">input_file</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">new_genome</span> <span class="ow">in</span> <span class="n">new_genomes</span><span class="p">:</span>
                <span class="n">output_q</span><span class="o">.</span><span class="n">put</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">failed_file_path</span> <span class="o">=</span> <span class="n">get_file_path</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">exception</span><span class="p">(</span><span class="s">&quot;Unable to read in data from &#39;{0}&#39;!&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">failed_file_path</span><span class="p">))</span>
            <span class="c"># If the genome on the output queue is just a string, its contents</span>
            <span class="c"># are the name of the file we failed to read in.</span>
            <span class="n">output_q</span><span class="o">.</span><span class="n">put</span><span class="p">(</span><span class="n">failed_file_path</span><span class="p">)</span>
        <span class="n">input_file</span> <span class="o">=</span> <span class="n">input_q</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>
    <span class="n">output_q</span><span class="o">.</span><span class="n">put</span><span class="p">(</span><span class="bp">None</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="parse_input_files"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.parse_input_files">[docs]</a><span class="k">def</span> <span class="nf">parse_input_files</span><span class="p">(</span><span class="n">input_files</span><span class="p">,</span> <span class="n">num_threads</span><span class="p">,</span> <span class="n">genomes</span><span class="p">,</span> <span class="n">min_coverage</span><span class="p">,</span> <span class="n">min_proportion</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Use a pool of worker threads to, in parallel, read in the input files.</span>
<span class="sd">    Populate the genome collection with the read-in data.</span>
<span class="sd">    This is the &quot;poison pill&quot; thread management algorithm, where threads are</span>
<span class="sd">    each given a &quot;you can stop now&quot; task once the actual queue of tasks is</span>
<span class="sd">    complete.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_files:</span>
<span class="sd">        num_threads:</span>
<span class="sd">        genomes:</span>
<span class="sd">        min_coverage:</span>
<span class="sd">        min_proportion:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c"># Lines below marked &quot;Single-thread&quot; can be uncommented and replace</span>
    <span class="c"># the lines marked &quot;Multi-thread&quot; to get single-thread behavior.</span>
    <span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Process</span><span class="p">,</span> <span class="n">Queue</span>  <span class="c"># Multi-thread</span>
    <span class="c"># from queue import Queue  # Single-thread</span>
    <span class="kn">from</span> <span class="nn">time</span> <span class="kn">import</span> <span class="n">sleep</span>

    <span class="n">input_q</span> <span class="o">=</span> <span class="n">Queue</span><span class="p">()</span>
    <span class="n">output_q</span> <span class="o">=</span> <span class="n">Queue</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">input_file</span> <span class="ow">in</span> <span class="n">input_files</span><span class="p">:</span>
        <span class="n">input_q</span><span class="o">.</span><span class="n">put</span><span class="p">(</span><span class="n">input_file</span><span class="p">)</span>
    <span class="c"># If the number of jobs is already smaller than the thread pool...</span>
    <span class="k">if</span> <span class="n">num_threads</span> <span class="o">&gt;</span> <span class="n">input_q</span><span class="o">.</span><span class="n">qsize</span><span class="p">():</span>
        <span class="n">num_threads</span> <span class="o">=</span> <span class="n">input_q</span><span class="o">.</span><span class="n">qsize</span><span class="p">()</span>
    <span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">thread_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">current_thread</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">num_threads</span><span class="p">):</span>
        <span class="n">input_q</span><span class="o">.</span><span class="n">put</span><span class="p">(</span><span class="bp">None</span><span class="p">)</span>
        <span class="n">current_thread</span> <span class="o">=</span> <span class="n">Process</span><span class="p">(</span><span class="n">target</span><span class="o">=</span><span class="n">manage_input_thread</span><span class="p">,</span>
                                 <span class="n">args</span><span class="o">=</span><span class="p">[</span><span class="n">genomes</span><span class="o">.</span><span class="n">reference</span><span class="p">(),</span> <span class="n">min_coverage</span><span class="p">,</span> <span class="n">min_proportion</span><span class="p">,</span> <span class="n">input_q</span><span class="p">,</span>
                                       <span class="n">output_q</span><span class="p">])</span>  <span class="c"># Multi-thread</span>
        <span class="n">current_thread</span><span class="o">.</span><span class="n">start</span><span class="p">()</span>  <span class="c"># Multi-thread</span>
        <span class="c">#manage_input_thread( genomes.reference(), min_coverage, min_proportion, input_q, output_q )  # Single-thread</span>
        <span class="n">thread_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_thread</span><span class="p">)</span>  <span class="c"># Multi-thread</span>
    <span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">while</span> <span class="n">num_threads</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">new_genome</span> <span class="o">=</span> <span class="n">output_q</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">new_genome</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
            <span class="n">num_threads</span> <span class="o">-=</span> <span class="mi">1</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">new_genome</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="c"># Reading this file in failed.  We only know the filename.</span>
            <span class="n">genomes</span><span class="o">.</span><span class="n">add_failed_genome</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">genomes</span><span class="o">.</span><span class="n">add_genome</span><span class="p">(</span><span class="n">new_genome</span><span class="p">)</span>
    <span class="n">sleep</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">current_thread</span> <span class="ow">in</span> <span class="n">thread_list</span><span class="p">:</span>  <span class="c"># Multi-thread</span>
        <span class="n">current_thread</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>  <span class="c"># Multi-thread</span>

</div>
<div class="viewcode-block" id="write_output_matrices"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.write_output_matrices">[docs]</a><span class="k">def</span> <span class="nf">write_output_matrices</span><span class="p">(</span><span class="n">genomes</span><span class="p">,</span> <span class="n">matrix_folder</span><span class="p">,</span> <span class="n">matrix_format_choices</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write matrices from genome collection data.</span>
<span class="sd">    Defines the matrix types for future expansion of custom matrix options.</span>
<span class="sd">    This information eventually should come from the user interface and be</span>
<span class="sd">    included in the XML configuration file, rather than hardcoded here.</span>
<span class="sd">    The matrix_format_choices option comes from the XML to here.</span>

<span class="sd">    Args:</span>
<span class="sd">        genomes:</span>
<span class="sd">        matrix_folder (str):</span>
<span class="sd">        matrix_format_choices:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">matrix_formats</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">{</span>
            <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/master_matrix.tsv&#39;</span><span class="p">,</span>
            <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;matrix&#39;</span><span class="p">,</span>
            <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;allcallable&#39;</span>
        <span class="p">},</span>
        <span class="p">{</span>
            <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/bestsnp_matrix.tsv&#39;</span><span class="p">,</span>
            <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;matrix&#39;</span><span class="p">,</span>
            <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;bestsnp&#39;</span>
        <span class="p">},</span>
        <span class="p">{</span>
            <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/bestsnp_matrix.snpfasta&#39;</span><span class="p">,</span>
            <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;fasta&#39;</span><span class="p">,</span>
            <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;bestsnp&#39;</span>
        <span class="p">},</span>
        <span class="p">{</span>
            <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/bestsnp_matrix.vcf&#39;</span><span class="p">,</span>
            <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;vcf&#39;</span><span class="p">,</span>
            <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;bestsnp&#39;</span>
        <span class="p">},</span>
        <span class="p">{</span>
            <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/missingdata_matrix.tsv&#39;</span><span class="p">,</span>
            <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;matrix&#39;</span><span class="p">,</span>
            <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;missingdata&#39;</span>
        <span class="p">},</span>
        <span class="p">{</span>
            <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/missingdata_matrix.snpfasta&#39;</span><span class="p">,</span>
            <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;fasta&#39;</span><span class="p">,</span>
            <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;missingdata&#39;</span>
        <span class="p">},</span>
        <span class="p">{</span>
            <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/missingdata_matrix.vcf&#39;</span><span class="p">,</span>
            <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;vcf&#39;</span><span class="p">,</span>
            <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;missingdata&#39;</span>
        <span class="p">}]</span>
    <span class="c"># The XML format transferring this information should be more elaborate someday.</span>
    <span class="k">if</span> <span class="n">matrix_format_choices</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="n">matrix_format_choices</span> <span class="o">==</span> <span class="s">&quot;include_allref_pos&quot;</span><span class="p">:</span>
        <span class="n">matrix_formats</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span> <span class="p">[</span>
            <span class="p">{</span>
                <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/withallrefpos_matrix.tsv&#39;</span><span class="p">,</span>
                <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;matrix&#39;</span><span class="p">,</span>
                <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;includeref&#39;</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/withallrefpos_matrix.snpfasta&#39;</span><span class="p">,</span>
                <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;fasta&#39;</span><span class="p">,</span>
                <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;includeref&#39;</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="s">&#39;filename&#39;</span><span class="p">:</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">matrix_folder</span> <span class="o">+</span> <span class="s">&#39;/withallrefpos_matrix.vcf&#39;</span><span class="p">,</span>
                <span class="s">&#39;dataformat&#39;</span><span class="p">:</span> <span class="s">&#39;vcf&#39;</span><span class="p">,</span>
                <span class="s">&#39;filter&#39;</span><span class="p">:</span> <span class="s">&#39;includeref&#39;</span>
            <span class="p">}]</span> <span class="p">)</span>
    <span class="n">genomes</span><span class="o">.</span><span class="n">write_to_matrices</span><span class="p">(</span><span class="n">matrix_formats</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="write_stats_data"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.write_stats_data">[docs]</a><span class="k">def</span> <span class="nf">write_stats_data</span><span class="p">(</span><span class="n">genomes</span><span class="p">,</span> <span class="n">stats_folder</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Write stats data from genome collection to preset filenames</span>

<span class="sd">    Args:</span>
<span class="sd">        genomes:</span>
<span class="sd">        stats_folder:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">general_stats_file</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">stats_folder</span> <span class="o">+</span> <span class="s">&#39;/general_stats.tsv&#39;</span>
    <span class="n">sample_stats_file</span> <span class="o">=</span> <span class="s">&#39;&#39;</span> <span class="o">+</span> <span class="n">stats_folder</span> <span class="o">+</span> <span class="s">&#39;/sample_stats.tsv&#39;</span>
    <span class="n">genomes</span><span class="o">.</span><span class="n">write_to_stats_files</span><span class="p">(</span><span class="n">general_stats_file</span><span class="p">,</span> <span class="n">sample_stats_file</span><span class="p">)</span>

</div>
<div class="viewcode-block" id="main"><a class="viewcode-back" href="../../nasp.html#nasp.vcf_to_matrix.main">[docs]</a><span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main flow control function for the script.  Is sequential over the</span>
<span class="sd">    following steps:</span>
<span class="sd">    1.  Parse arguments</span>
<span class="sd">    2.  Read in reference</span>
<span class="sd">    3.  Read in all query genomes in parallel</span>
<span class="sd">    4.  Write output matrices</span>
<span class="sd">    5.  Write stats files</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">commandline_args</span> <span class="o">=</span> <span class="n">_parse_args</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">dto_file</span><span class="p">:</span>
        <span class="n">commandline_args</span> <span class="o">=</span> <span class="n">_parse_input_config</span><span class="p">(</span><span class="n">commandline_args</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">WARNING</span><span class="p">)</span>
    <span class="kn">from</span> <span class="nn">nasp.nasp_objects</span> <span class="kn">import</span> <span class="n">ReferenceGenome</span><span class="p">,</span> <span class="n">GenomeCollection</span>

    <span class="n">reference</span> <span class="o">=</span> <span class="n">ReferenceGenome</span><span class="p">()</span>
    <span class="n">import_reference</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">reference_fasta</span><span class="p">,</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">reference_dups</span><span class="p">)</span>
    <span class="n">genomes</span> <span class="o">=</span> <span class="n">GenomeCollection</span><span class="p">()</span>
    <span class="n">genomes</span><span class="o">.</span><span class="n">set_reference</span><span class="p">(</span><span class="n">reference</span><span class="p">)</span>
    <span class="n">parse_input_files</span><span class="p">(</span><span class="n">commandline_args</span><span class="o">.</span><span class="n">input_files</span><span class="p">,</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">num_threads</span><span class="p">,</span> <span class="n">genomes</span><span class="p">,</span>
                      <span class="n">commandline_args</span><span class="o">.</span><span class="n">minimum_coverage</span><span class="p">,</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">minimum_proportion</span><span class="p">)</span>
    <span class="n">write_output_matrices</span><span class="p">(</span><span class="n">genomes</span><span class="p">,</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">matrix_folder</span><span class="p">,</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">filter_matrix_format</span><span class="p">)</span>
    <span class="n">write_stats_data</span><span class="p">(</span><span class="n">genomes</span><span class="p">,</span> <span class="n">commandline_args</span><span class="o">.</span><span class="n">stats_folder</span><span class="p">)</span>

</div>
<span class="k">if</span> <span class="n">__name__</span> <span class="o">==</span> <span class="s">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, Darrin Lemmer.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'0.9.8',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>