

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Usage &mdash; Northern Arizona SNP Pipeline 1.0.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="Northern Arizona SNP Pipeline 1.0.0 documentation" href="index.html"/>
        <link rel="next" title="Contributing" href="contributing.html"/>
        <link rel="prev" title="Installation" href="install.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Northern Arizona SNP Pipeline
          

          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#nasp">nasp</a></li>
<li class="toctree-l2"><a class="reference internal" href="#format-fasta">format_fasta</a></li>
<li class="toctree-l2"><a class="reference internal" href="#find-duplicates">find_duplicates</a></li>
<li class="toctree-l2"><a class="reference internal" href="#vcf-to-matrix">vcf_to_matrix</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#matrices">Matrices</a></li>
<li class="toctree-l3"><a class="reference internal" href="#statistics">Statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example-statistics">Example Statistics</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">Academic and Research License</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Northern Arizona SNP Pipeline</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Usage</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/usage.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>The <a class="reference internal" href="#nasp">nasp</a> program is the pipeline. While you can run the other programs listed in this section directly, they are
primarily internal functions of the nasp pipeline.</p>
<div class="section" id="nasp">
<span id="id1"></span><h2>nasp<a class="headerlink" href="#nasp" title="Permalink to this headline">¶</a></h2>
<p>The first time you run <cite>nasp</cite> will typically be without arguments where it will ask you where your files are located and
what kinds of analyses you want to run.:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">nasp</span>
</pre></div>
</div>
<p>These details will be saved to a [run-name]-config.xml file in the output folder.
After the first run, you can use this configuration to repeat or modify the analysis without going through the
prompts.:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">nasp</span> <span class="o">--</span><span class="n">config</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="p">[</span><span class="n">run</span><span class="o">-</span><span class="n">name</span><span class="p">]</span><span class="o">-</span><span class="n">config</span><span class="o">.</span><span class="n">xml</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>usage: nasp [-h] [--config CONFIG] [reference_fasta] [output_folder]
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="39%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Positional Arguments</th>
<th class="head">(Optional)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ReferenceFasta</td>
<td>Path to the reference fasta.</td>
</tr>
<tr class="row-odd"><td>OutputFolder</td>
<td>Folder to store the output files.</td>
</tr>
</tbody>
</table>
<table border="1" class="docutils">
<colgroup>
<col width="39%" />
<col width="61%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Flags</th>
<th class="head">(Optional)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>-h, &#8211;help</td>
<td>show this help message and exit</td>
</tr>
<tr class="row-odd"><td>&#8211;config CONFIG</td>
<td>Path to the configuration xml file.</td>
</tr>
</tbody>
</table>
<p>The following are examples of the nasp commandline prompts:</p>
<div class="highlight-python"><div class="highlight"><pre>$ nasp
Welcome to NASP version 1.0.0.
</pre></div>
</div>
<p>NASP will write all the files it creates including any vcf, bam, fasta, tsv, etc to this output folder:</p>
<div class="highlight-python"><div class="highlight"><pre>Where would you like output files to be written [nasp_results]? results

Where is the reference fasta file you would like to use? ./examples/example_1/reference.fasta
</pre></div>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">When entering file paths, try typing the first few characters followed by the <cite>Tab</cite> key to complete the folder name.
Tab completion is only supported if the Python <cite>readline</cite> module is available.</p>
</div>
<p>The nucmer aligner will scan the reference file for duplicate regions creating duplicates.txt and reference.delta in the
output reference/ folder.:</p>
<div class="highlight-python"><div class="highlight"><pre>Do you want to check the reference for duplicated regions
and skip SNPs that fall in those regions [Y]?
</pre></div>
</div>
<p>If you select <cite>none</cite> for the job manager, nasp will run the analysis in the background and attempt to balance the job
load. If possible, it is better to use a job manager as the analysis can take a long time consuming a lot of computer resources. The
defaults queue and commandline arguments should be sufficient.:</p>
<div class="highlight-python"><div class="highlight"><pre>What system do you use for job management (PBS/TORQUE, SLURM, SGE/OGE, and &#39;none&#39; are currently supported) [PBS]?
  Would you like to specify a queue/partition to use for all jobs (leave blank to use default queue) []?
  What additional arguments do you need to pass to the job management system []?
</pre></div>
</div>
<p>External genomes are fasta files not created by NASP such as from a public repository. The NUCmer aligner will be used
to align the genomes against the reference genome.:</p>
<div class="highlight-python"><div class="highlight"><pre>Do you have fasta files for external genomes you wish to include [Y]?
Where are these files located [/home/tgen/NASP]? ./examples/example_1
  Would you like to set advanced NUCmer settings [N]?
</pre></div>
</div>
<p>fastq.gz read files:</p>
<div class="highlight-python"><div class="highlight"><pre>Do you have read files you wish to include [Y]?
Where are these files located [/home/tgen/NASP]? ./examples/example_1

Would you like to use Trimmomatic to trim your reads first [N]? Y
    What adapter file are you using for trimming?
    Would you also like to perform quality trimming [N]? Y
    What quality trimming parameters do you want to use [SLIDINGWINDOW:5:20]?
    What is the minimum length read to keep after trimming [80]?
Would you like to set advanced ReadTrimmer settings [N]?
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>This pipeline currently supports three aligners: BWA, Novoalign, and SNAP.
You can also provide pre-aligned BAM files, and you can choose as many options as you want.

Would you like to run BWA samp/se [N]?*

Would you like to run BWA mem [Y]?
Would you like to set advanced BWA-mem settings [N]?

Would you like to run Bowtie2 [Y]?
Would you like to set advanced Bowtie2 settings [N]?

Would you like to run Novoalign [Y]?
Would you like to set advanced Novoalign settings [N]?

Would you like to run SNAP [N]?*

Do you have pre-aligned SAM/BAM files you wish to include [N]?
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>This pipeline currently supports four SNP callers: GATK, SolSNP, VarScan, and SAMtools, and you can provide VCF files.
You can choose as many options as you want.

Would you like to run GATK [Y]?

Unable to find &#39;GenomeAnalysisTK.jar&#39;, please enter the full path to &#39;GenomeAnalysisTK.jar&#39;: /packages/GenomeAnalysisTK/2.7-2/GenomeAnalysisTK.jar
Would you like to set advanced GATK settings [N]?

Would you like to run SolSNP [N]?

Would you like to run VarScan [Y]?
Would you like to set advanced VarScan settings [N]?

Would you like to run SAMtools [Y]?
Would you like to set advanced SAMtools settings [N]?

Unable to find &#39;CreateSequenceDictionary.jar&#39;, please enter the full path to &#39;CreateSequenceDictionary.jar&#39;: /packages/tnorth/bin/CreateSequenceDictionary.jar
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>Do you have pre-called VCFfiles you wish to include [N]?
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre>This pipeline can do filtering based on coverage.
If you do not want filtering based on coverage, enter 0.
What is your minimum coverage threshold [10]?

This pipeline can do filtering based on the proportion of reads that match the call made by the SNP caller.
If you do not want filtering based on proportion, enter 0.
What is the minimum acceptable proportion [0.9]?
</pre></div>
</div>
<p>See <a class="reference internal" href="#vcf-to-matrix">vcf_to_matrix</a> for commandline arguments you can pass to the MatrixGenerator. This is not typically required.:</p>
<div class="highlight-python"><div class="highlight"><pre>Would you like to set advanced MatrixGenerator settings [N]?
</pre></div>
</div>
<p>In addition to the statistics, bestsnp, missing data, and master matrices, vcf_to_matrix, will create withallrefpos
matrices in the output <cite>matrices/</cite> folder. See <a class="reference internal" href="#vcf-to-matrix">vcf_to_matrix</a> for output details.</p>
<div class="highlight-python" id="create-withallrefpos-matrix-prompt"><div class="highlight"><pre>Do you want to create a matrix that includes all reference positions, but with low-quality calls masked [N]?
</pre></div>
</div>
</div>
<div class="section" id="format-fasta">
<h2>format_fasta<a class="headerlink" href="#format-fasta" title="Permalink to this headline">¶</a></h2>
<p>Reformats a fasta to be split 80 characters per line, with system line-endings.:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: format_fasta [-h] --inputfasta INPUTFASTA --outputfasta OUTPUTFASTA
</pre></div>
</div>
<p>Options:</p>
<blockquote>
<div><table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></td>
<td>show this help message and exit</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--inputfasta <var>file</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to input fasta.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--outputfasta <var>file</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to output fasta.</td></tr>
</tbody>
</table>
</div></blockquote>
</div>
<div class="section" id="find-duplicates">
<h2>find_duplicates<a class="headerlink" href="#find-duplicates" title="Permalink to this headline">¶</a></h2>
<p>Scans the reference genome for duplicate regions using the NUCmer aligner.:</p>
<div class="highlight-python"><div class="highlight"><pre>usage: find_duplicates [-h] [--nucmerpath NUCMERPATH] --reference REFERENCE
</pre></div>
</div>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-h</span>, <span class="option">--help</span></kbd></td>
<td>show this help message and exit</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--nucmerpath <var>NUCMERPATH</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to the &#8216;nucmer&#8217; executable.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--reference <var>REFERENCE</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to the reference fasta file.</td></tr>
</tbody>
</table>
</div>
<div class="section" id="vcf-to-matrix">
<span id="id2"></span><h2>vcf_to_matrix<a class="headerlink" href="#vcf-to-matrix" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><div class="highlight"><pre>vcf_to_matrix --mode=xml --num-threads=1 --dto-file DTO_FILE
</pre></div>
</div>
<p>Options:</p>
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--mode <var>mode</var></span></kbd></td>
<td>Data passing mode
MODE  can be either &#8216;commandline&#8217; or &#8216;xml&#8217;.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--reference-fasta <var>file</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to input reference fasta file.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--reference-dups <var>file</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to input reference dups file.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--input-files <var>files</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to input VCF/fasta files for matrix conversion.
INPUT_FILES is a list of file paths</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--matrix-folder <var>path</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Name of folder to write output matries to.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--stats-folder <var>path</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Name of folder to write statistics files to.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minimum-coverage <var>number</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Minimum coverage depth at a position.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minimum-proportion <var>number</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Minimum proportion of reads that must match the call at a position.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--num-threads <var>number</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Number of threads to use when processing input.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--dto-file <var>file</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to a matrix_dto XML file that defines all the parameters.</td></tr>
</tbody>
</table>
<div class="section" id="matrices">
<h3>Matrices<a class="headerlink" href="#matrices" title="Permalink to this headline">¶</a></h3>
<p><a class="reference internal" href="#vcf-to-matrix">vcf_to_matrix</a> will write the following matrices to the output matrices/ folder in tsv, snpfasta, and vcf formats:</p>
<table border="1" class="docutils">
<colgroup>
<col width="24%" />
<col width="76%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Matrix</th>
<th class="head">Meaning</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Master Matrix</td>
<td>All positions</td>
</tr>
<tr class="row-odd"><td>Withallrefpos</td>
<td>Positions that passed the <a class="reference internal" href="#general-stats">General Stats</a> quality_breadth filter <a href="#id16"><span class="problematic" id="id17"><span id="id3"></span>[3]_</span></a></td>
</tr>
<tr class="row-even"><td>Best SNP</td>
<td>SNPs that passed the <a class="reference internal" href="#general-stats">General Stats</a> quality_breadth filter</td>
</tr>
<tr class="row-odd"><td>Missing Data</td>
<td>Positions that passed the <a class="reference internal" href="#general-stats">General Stats</a> quality_breadth filter</td>
</tr>
</tbody>
</table>
<a class="reference internal image-reference" href="_images/MatricesVennDiagram.png"><img alt="vcf_to_matrix output matrices venn diagram" class="align-right" src="_images/MatricesVennDiagram.png" style="width: 50%;" /></a>
<dl class="docutils">
<dt>The conventions used for what data is stored are as follows:</dt>
<dd><dl class="first last docutils">
<dt>Genomes:</dt>
<dd><ul class="first last simple">
<li>A, C, G, T, U:  The respective call.</li>
<li>N:  Called &#8220;N&#8221; according to upstream analysis tools.</li>
<li>X:  Not called by upstream analysis tools.</li>
<li>. or empty string:  A deletion relative to reference.</li>
<li>String of length &gt;1:  An insertion relative to reference.</li>
<li>Any other single letter:  A degeneracy.</li>
</ul>
</dd>
<dt>Duplicate region data:</dt>
<dd><ul class="first last simple">
<li>0:  Position not in a region that is duplicated within the reference.</li>
<li>1:  Position is in a region that is duplicated.</li>
<li>-:  Duplicate checking at this position was skipped by the user.</li>
</ul>
</dd>
<dt>Filters:</dt>
<dd><ul class="first last simple">
<li>Y:  This position passed its filter.</li>
<li>N:  This position failed its filter.</li>
<li>?:  The filter could not be checked, and so the position is assumed to have failed.</li>
<li><cite>-</cite>:  The filter was not applicable, or skipped, or could not be checked for a known reason, and so is assumed to have passed.</li>
</ul>
</dd>
</dl>
</dd>
</dl>
</div>
<div class="section" id="statistics">
<h3>Statistics<a class="headerlink" href="#statistics" title="Permalink to this headline">¶</a></h3>
<p><a class="reference internal" href="#vcf-to-matrix">vcf_to_matrix</a> collects sample analysis statistics and stores them as TSV files in the output statistics/ folder.
The tables below list and describe their columns.</p>
<p>General Stats include statistics gathered for all samples relative to the reference genome.</p>
<table border="1" class="docutils" id="general-stats">
<colgroup>
<col width="27%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">general_stats.tsv</th>
<th class="head">Descriptions</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Contig</td>
<td>The contig name defined by its source file.</td>
</tr>
<tr class="row-odd"><td>reference_length</td>
<td>Total number of positions found in the the reference genome.</td>
</tr>
<tr class="row-even"><td>reference_clean</td>
<td>Number of positions called A/C/G/T in the reference genome.</td>
</tr>
<tr class="row-odd"><td>reference_clean (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>reference_duplicated</td>
<td>Number of reference contig positions in a duplicated region.</td>
</tr>
<tr class="row-odd"><td>reference_duplicated (%)</td>
<td>Percent of the reference contig</td>
</tr>
<tr class="row-even"><td>all_called</td>
<td>Number of positions where the base was called A/C/G/T in all samples.</td>
</tr>
<tr class="row-odd"><td>all_called (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>all_passed_coverage</td>
<td>Number of positions that passed the coverage filter in all samples. <a class="footnote-reference" href="#id11" id="id5">[1]</a></td>
</tr>
<tr class="row-odd"><td>all_passed_coverage (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>all_passed_proportion</td>
<td>Maximum number of positions that passed the proportion filter in all samples. <a class="footnote-reference" href="#id11" id="id6">[1]</a></td>
</tr>
<tr class="row-odd"><td>all_passed_proportion (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>all_passed_consensus</td>
<td>Number of positions where all analyses agreed for all samples.</td>
</tr>
<tr class="row-odd"><td>all_passed_consensus (%)</td>
<td>Percent of positions where all samples matched the reference contig.</td>
</tr>
<tr class="row-even"><td>quality_breadth</td>
<td>Number of positions called A/C/G/T and passed all filters for all samples.</td>
</tr>
<tr class="row-odd"><td>quality_breadth (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>any_snps</td>
<td>Number of positions that had a SNP called A/C/G/T in any sample.</td>
</tr>
<tr class="row-odd"><td>any_snps (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>best_snps</td>
<td>Number of positions that had a confident SNP called A/C/G/T in any sample.</td>
</tr>
<tr class="row-odd"><td>best_snps (%)</td>
<td>Percentage of above.</td>
</tr>
</tbody>
</table>
<p>Sample Stats include statistics for each sample/analysis combination.</p>
<table border="1" class="docutils">
<colgroup>
<col width="27%" />
<col width="73%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">sample_stats.tsv</th>
<th class="head">Column Descriptions</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Sample</td>
<td>Sample name based on the filename.</td>
</tr>
<tr class="row-odd"><td>Sample::Analysis</td>
<td>Sample/Aligner/SNP Caller combination.</td>
</tr>
<tr class="row-even"><td>was_called</td>
<td>Number of positions called A/C/G/T.</td>
</tr>
<tr class="row-odd"><td>was_called (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>passed_coverage_filter</td>
<td>Number of positions that passed the coverage filter. <a class="footnote-reference" href="#id11" id="id7">[1]</a></td>
</tr>
<tr class="row-odd"><td>passed_coverage_filter (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>passed_proportion_filter</td>
<td>Number of positions that passed the proportion filter. <a class="footnote-reference" href="#id11" id="id8">[1]</a></td>
</tr>
<tr class="row-odd"><td>passed_proportion_filter (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>quality_breadth</td>
<td>Number of positions called A/C/G/T and passed all filters. <a class="footnote-reference" href="#id11" id="id9">[1]</a></td>
</tr>
<tr class="row-odd"><td>quality_breadth (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>called_reference</td>
<td>Number of positions that matched the reference.</td>
</tr>
<tr class="row-odd"><td>called_reference (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>called_snp</td>
<td>Number of positions that differed from the reference.</td>
</tr>
<tr class="row-odd"><td>called_snp (%)</td>
<td>Percentage of above.</td>
</tr>
<tr class="row-even"><td>called_dgen</td>
<td>Number of positions not called A/C/G/T. <a class="footnote-reference" href="#id12" id="id10">[2]</a></td>
</tr>
<tr class="row-odd"><td>called_dgen (%)</td>
<td>Percentage of above.</td>
</tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id11" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[1]</td><td><em>(<a class="fn-backref" href="#id5">1</a>, <a class="fn-backref" href="#id6">2</a>, <a class="fn-backref" href="#id7">3</a>, <a class="fn-backref" href="#id8">4</a>, <a class="fn-backref" href="#id9">5</a>)</em> If the filter could not be checked for a known reason, such as with a FASTA file, it is assumed to have passed.</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="id12" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id10">[2]</a></td><td>Includes degeneracies, unknown, and uncalled</td></tr>
</tbody>
</table>
<p>The pseudo-flowcharts below reflect relationships between the statistics where each terminal node is a statistics column.
Click the image to view it in detail.</p>
<a class="reference internal image-reference" href="_images/StatisticsFlowchart.png"><img alt="Relationship flowchart between the statistics" src="_images/StatisticsFlowchart.png" style="width: 99%;" /></a>
<a class="reference internal image-reference" href="_images/StatisticsFlowchartFilters.png"><img alt="Relationship flowchart between the statistics filters" src="_images/StatisticsFlowchartFilters.png" style="width: 99%;" /></a>
</div>
<div class="section" id="example-statistics">
<h3>Example Statistics<a class="headerlink" href="#example-statistics" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils" id="id13">
<caption><span class="caption-text">general_stats.tsv</span><a class="headerlink" href="#id13" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
<col width="5%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Contig</th>
<th class="head">reference_length</th>
<th class="head">reference_clean</th>
<th class="head">reference_clean (%)</th>
<th class="head">reference_duplicated</th>
<th class="head">reference_duplicated (%)</th>
<th class="head">all_called</th>
<th class="head">all_called (%)</th>
<th class="head">all_passed_coverage</th>
<th class="head">all_passed_coverage (%)</th>
<th class="head">all_passed_proportion</th>
<th class="head">all_passed_proportion (%)</th>
<th class="head">all_passed_consensus</th>
<th class="head">all_passed_consensus (%)</th>
<th class="head">quality_breadth</th>
<th class="head">quality_breadth (%)</th>
<th class="head">any_snps</th>
<th class="head">any_snps (%)</th>
<th class="head">best_snps</th>
<th class="head">best_snps (%)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Whole Genome</td>
<td>3977</td>
<td>3977</td>
<td>100.00%</td>
<td>0</td>
<td>0.00%</td>
<td>3974</td>
<td>99.92%</td>
<td>3974</td>
<td>99.92%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>5</td>
<td>0.13%</td>
<td>5</td>
<td>0.13%</td>
</tr>
<tr class="row-odd"><td>500WT1_test</td>
<td>3977</td>
<td>3977</td>
<td>100.00%</td>
<td>0</td>
<td>0.00%</td>
<td>3974</td>
<td>99.92%</td>
<td>3974</td>
<td>99.92%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>5</td>
<td>0.13%</td>
<td>5</td>
<td>0.13%</td>
</tr>
</tbody>
</table>
<p>The [any] and [all] rows track statistics collected for any and all analysis combinations on each sample. The first two
[any] and [all] rows are special because they track statistics collected for any and all analysis combinations for all
samples.</p>
<table border="1" class="docutils" id="id14">
<caption><span class="caption-text">sample_stats.tsv</span><a class="headerlink" href="#id14" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
<col width="6%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Sample</th>
<th class="head">Sample::Analysis</th>
<th class="head">was_called</th>
<th class="head">was_called (%)</th>
<th class="head">passed_coverage_filter</th>
<th class="head">passed_coverage_filter (%)</th>
<th class="head">passed_proportion_filter</th>
<th class="head">passed_proportion_filter (%)</th>
<th class="head">quality_breadth</th>
<th class="head">quality_breadth (%)</th>
<th class="head">called_reference</th>
<th class="head">called_reference (%)</th>
<th class="head">called_snp</th>
<th class="head">called_snp (%)</th>
<th class="head">called_degen</th>
<th class="head">called_degen (%)</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>[any]</td>
<td>&nbsp;</td>
<td>3974</td>
<td>99.92%</td>
<td>3974</td>
<td>99.92%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>3968</td>
<td>99.77%</td>
<td>5</td>
<td>0.13%</td>
<td>0</td>
<td>0.00%</td>
</tr>
<tr class="row-odd"><td>[all]</td>
<td>&nbsp;</td>
<td>3974</td>
<td>99.92%</td>
<td>3974</td>
<td>99.92%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>3968</td>
<td>99.77%</td>
<td>5</td>
<td>0.13%</td>
<td>0</td>
<td>0.00%</td>
</tr>
<tr class="row-even"><td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
<td>&nbsp;</td>
</tr>
<tr class="row-odd"><td>example_1_L001</td>
<td>[any]</td>
<td>3974</td>
<td>99.92%</td>
<td>3974</td>
<td>99.92%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>3968</td>
<td>99.77%</td>
<td>5</td>
<td>0.13%</td>
<td>0</td>
<td>0.00%</td>
</tr>
<tr class="row-even"><td>example_1_L001</td>
<td>[all]</td>
<td>3974</td>
<td>99.92%</td>
<td>3974</td>
<td>99.92%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>3968</td>
<td>99.77%</td>
<td>5</td>
<td>0.13%</td>
<td>0</td>
<td>0.00%</td>
</tr>
<tr class="row-odd"><td>example_1_L001</td>
<td>example_1_L001::Bowtie2,GATK</td>
<td>3974</td>
<td>99.92%</td>
<td>3974</td>
<td>99.92%</td>
<td>3973</td>
<td>99.90%</td>
<td>3973</td>
<td>99.90%</td>
<td>3968</td>
<td>99.77%</td>
<td>5</td>
<td>0.13%</td>
<td>0</td>
<td>0.00%</td>
</tr>
</tbody>
</table>
<p>Running NASP with the example data and default options on a PBS cluster results in the following output.:</p>
<div class="highlight-python"><div class="highlight"><pre>nasp_results
├── bowtie2
│   ├── example_1_L001-bowtie2.bam
│   ├── example_1_L001-bowtie2.bam.bai
│   ├── example_1_L001-bowtie2.mpileup
│   ├── nasp_bowtie2_example_1_L001.e293242
│   └── nasp_bowtie2_example_1_L001.o293242
├── bwamem
│   ├── example_1_L001-bwamem.bam
│   ├── example_1_L001-bwamem.bam.bai
│   ├── example_1_L001-bwamem.mpileup
│   ├── nasp_bwamem_example_1_L001.e293241
│   └── nasp_bwamem_example_1_L001.o293241
├── external
│   ├── example_1.delta
│   ├── example_1.fasta
│   ├── example_1.filtered.delta
│   ├── example_1.frankenfasta
│   ├── nasp_AssemblyImporter_example_1.e293240
│   └── nasp_AssemblyImporter_example_1.o293240
├── gatk
│   ├── example_1_L001-bowtie2-gatk.vcf
│   ├── example_1_L001-bowtie2-gatk.vcf.idx
│   ├── example_1_L001-bwamem-gatk.vcf
│   ├── example_1_L001-bwamem-gatk.vcf.idx
│   ├── example_1_L001-novo-gatk.vcf
│   ├── example_1_L001-novo-gatk.vcf.idx
│   ├── nasp_gatk_example_1_L001-bowtie2.e293247
│   ├── nasp_gatk_example_1_L001-bowtie2.o293247
│   ├── nasp_gatk_example_1_L001-bwamem.e293244
│   ├── nasp_gatk_example_1_L001-bwamem.o293244
│   ├── nasp_gatk_example_1_L001-novo.e293250
│   └── nasp_gatk_example_1_L001-novo.o293250
├── matrices
│   ├── bestsnp_matrix.snpfasta
│   ├── bestsnp_matrix.tsv
│   ├── bestsnp_matrix.vcf
│   ├── master_matrix.tsv
│   ├── missingdata_matrix.snpfasta
│   ├── missingdata_matrix.tsv
│   └── missingdata_matrix.vcf
├── matrix_dto.xml
├── nasp_matrix.e293253
├── nasp_matrix.o293253
├── novo
│   ├── example_1_L001-novo.bam
│   ├── example_1_L001-novo.bam.bai
│   ├── example_1_L001-novo.mpileup
│   ├── nasp_novo_example_1_L001.e293243
│   └── nasp_novo_example_1_L001.o293243
├── reference
│   ├── duplicates.txt
│   ├── nasp_DupFinder.e293239
│   ├── nasp_DupFinder.o293239
│   ├── nasp_index.e293238
│   ├── nasp_index.o293238
│   ├── reference.1.bt2
│   ├── reference.2.bt2
│   ├── reference.3.bt2
│   ├── reference.4.bt2
│   ├── reference.delta
│   ├── reference.dict
│   ├── reference.fasta
│   ├── reference.fasta.amb
│   ├── reference.fasta.ann
│   ├── reference.fasta.bwt
│   ├── reference.fasta.fai
│   ├── reference.fasta.idx
│   ├── reference.fasta.pac
│   ├── reference.fasta.sa
│   ├── reference.rev.1.bt2
│   └── reference.rev.2.bt2
├── nasp_results-config.xml
├── runlog.txt
├── samtools
│   ├── example_1_L001-bowtie2-samtools.vcf
│   ├── example_1_L001-bwamem-samtools.vcf
│   ├── example_1_L001-novo-samtools.vcf
│   ├── nasp_samtools_example_1_L001-bowtie2.e293249
│   ├── nasp_samtools_example_1_L001-bowtie2.o293249
│   ├── nasp_samtools_example_1_L001-bwamem.e293246
│   ├── nasp_samtools_example_1_L001-bwamem.o293246
│   ├── nasp_samtools_example_1_L001-novo.e293252
│   └── nasp_samtools_example_1_L001-novo.o293252
├── statistics
│   ├── general_stats.tsv
│   └── sample_stats.tsv
└── varscan
    ├── example_1_L001-bowtie2-varscan.vcf
    ├── example_1_L001-bowtie2.txt
    ├── example_1_L001-bwamem-varscan.vcf
    ├── example_1_L001-bwamem.txt
    ├── example_1_L001-novo-varscan.vcf
    ├── example_1_L001-novo.txt
    ├── nasp_varscan_example_1_L001-bowtie2.e293248
    ├── nasp_varscan_example_1_L001-bowtie2.o293248
    ├── nasp_varscan_example_1_L001-bwamem.e293245
    ├── nasp_varscan_example_1_L001-bwamem.o293245
    ├── nasp_varscan_example_1_L001-novo.e293251
    └── nasp_varscan_example_1_L001-novo.o293251
</pre></div>
</div>
<p>Most of the files are either from the external analysis programs or STDIN and STDOUT files created by the PBS cluster
for each job. The files in this example created directly by NASP include:</p>
<ul class="simple">
<li>The matrices and statistics folders from <a class="reference internal" href="#vcf-to-matrix">vcf_to_matrix</a></li>
<li>runlog.txt which includes the terminal commands used to run the external analysis programs</li>
<li>nasp_results-config.xml</li>
<li>matrix_dto.xml</li>
<li>reference/duplicates.txt</li>
<li>external/example_1.frankenfasta</li>
</ul>
<p>The other files are organized into folders based on the analysis tool used to create them.</p>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="contributing.html" class="btn btn-neutral float-right" title="Contributing" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="install.html" class="btn btn-neutral" title="Installation" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, TGen North.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>